{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Project\n",
    "\n",
    "The file `mRNA.fasta` contains a set of mRNA sequences expressed by a hypothetical organism. \n",
    " \n",
    "The objective of this project is to analyse these target sequences to discover what proteins the organism produces and how they differ from the corresponding proteins in other species."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. Coding sequence extraction\n",
    "\n",
    "Extracts the coding sequences from the mRNA sequences.\n",
    "\n",
    "a) For each target mRNA sequence, finds the longest open reading frame (ORF) in the sequence. An open reading frame is a continuous sequence that has the start codon `ATG` in the beginning and a stop codon at the end. Remember that the start codon can be at any position in the sequence.\n",
    "\n",
    "b) Translates the ORFs to protein sequences for further analysis.\n",
    "\n",
    "c) Saves the target protein sequences to the file `1-proteins.gb` in the GenBank format.\n",
    "\n",
    "d) Summarises discoveries by reporting the position and length of the longest ORF for each target mRNA sequence.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence name: mRNA_1\n",
      "Protein name: mRNA_1\n",
      "Protein lenght: 616\n",
      "The starting position of ORF in RNA sequence: 437\n",
      "The ending position of ORF in RNA sequence: 2285\n",
      "Translated protein: MSGKAQQQSRIKELIVLGREQGYLTYAEVNDHLPEDISDPEQVEDIIRMINDMGINVFEAAPDKDSLMLADADTDEAAAEEAAAALAAVETDIGRTTDPVRMYMREMGTVELLTREGEIEIAKRIEEGIREVMGAIAHFPGTVDHILSEYTRVTTEGGRLSDVLSGYIDPDDGIAPPAAEVPPPVGAKTAKADDDTDDDEAEASSDDEEEVESGPDPIIAAQRFGAVSDQMEITRKALKKHGRGNKQAIAELIALAELFMPIKLVPKQFEGLVERVRSALERLRAQERAIMQLCVRDARMPRADFVRQFPGNEVDESWTDALAKGKAKYAEAIGRLQPDIIRCQQKLTALETETGLTIAEIKDINRRMSIGEAKARRAKKEMVEANLRLVISIAKKYTNRGLQFLDLIQEGNIGLMKAVDKFEYRRGYKFSTYATWWIRQAITRSIADQARTIRIPVHMIETINKLNRISRQMLQEMGREPTPEELGERMEMPEDKIRKVLKIAKEPISMETPIGDDEDSHLGDFIEDSTMQSPIDVATVESLKEATRDVLSGLTAREAKVLRMRFGIDMNTDHTLEEVGKQFDVTRERIRQIEAKALRKLRHPTRSEHLRSFLDE\n",
      "\n",
      "Sequence name: mRNA_2\n",
      "Protein name: mRNA_2\n",
      "Protein lenght: 372\n",
      "The starting position of ORF in RNA sequence: 42\n",
      "The ending position of ORF in RNA sequence: 1158\n",
      "Translated protein: MTVTPERSTYTPPGRLGDESSGPRTDPRFSPAMVEALATFGLDEVAAAPPVSASDDLPTVLAAVGASHDGFQAVYDSIALDLPTDRDDVETSTETILGVDGNEITLHVFRPAGVEGVLPGLVYTHGGGMTILTTDNRVHRRWCTDLAAAGSVVVMVDFRNAWTAEGHHPFPSGVEDCLAAVLWVDEHRESLGLSGVVVQGESGGGNLAIATTLLAKRRGRLDAIDGVYASIPYISGGYAWDHERRLTELPSLVENDGYFIENGGMALLVRAYDPTGEHAEDPIAWPYFASEDELRGLPPFVVAVNELDPLRDEGIAFARRLARAGVDVAARVNIGLVHGADVIFRHWLPAALESTVRDVAGFAADRARLRRP\n",
      "\n",
      "Sequence name: mRNA_3\n",
      "Protein name: mRNA_3\n",
      "Protein lenght: 348\n",
      "The starting position of ORF in RNA sequence: 0\n",
      "The ending position of ORF in RNA sequence: 1044\n",
      "Translated protein: MKGLILAAGRGTRLRPLTHTRPKPAIRVAGRPIIHYAVENLLEAGVREIGVVVSPETERDLKVALEGYPVRYVFQEEPQGLAHAVDVARGFLGESPFVLYLGDNLFQKGIRRFLEAFKPGVSAVIALVRVEDPRQFGVAVLEGNRVVRLLEKPKEPPSDLAVAGVYVFSPEVLEVVRGLKPSARGEYEITDAIQGLIDRGKTVVGVEVEGWWKDTGRHQDLLDANRLLLEELTPKVEGDVVESQLTGRVVVEKGAKVVRSTVIGPAHIGAGAVVEEAFVGPFTSVGPRARVARSEVEYSLLEEEAVVEEVRLRLQECILGVRAEVKSRDGLPRAHRLILGDLSQVELA\n",
      "\n",
      "Sequence name: mRNA_4\n",
      "Protein name: mRNA_4\n",
      "Protein lenght: 995\n",
      "The starting position of ORF in RNA sequence: 186\n",
      "The ending position of ORF in RNA sequence: 3171\n",
      "Translated protein: MIVVGMGQDKEAAASEQKTTTVEENGNSATDNKTSETQTTATNVNHIEETQSYNATVTEQPSNATQVTTEEAPKAVQAPQTAQPANIETVKEEVVKEEAKPQVKETTQSQDNSGDQRQVDLTPKKATQNQVAETQVEVAQPRTASESKPRVTRSADVAEAKEASNAKVETGTDVTSKVTVEIGSIEGHNNTNKVEPHAGQRAVLKYKLKFENGLHQGDYFDFTLSNNVNTHGVSTARKVPEIKNGSVVMATGEVLEGGKIRYTFTNDIEDKVDVTAELEINLFIDPKTVQTNGNQTITSTLNEEQTSKELDVKYKDGIGNYYANLNGSIETFNKANNRFSHVAFIKPNNGKTTSVTVTGTLMKGSNQNGNQPKVRIFEYLGNNEDIAKSVYANTTDTSKFKEVTSNMSGNLNLQNNGSYSLNIENLDKTYVVHYDGEYLNGTDEVDFRTQMVGHPEQLYKYYYDRGYTLTWDNGLVLYSNKANGNEKNGPIIQNNKFEYKEDTIKETLTGQYDKNLVTTVEEEYDSSTLDIDYHTAIDGGGGYVDGYIETIEETDSSAIDIDYHTAVDSEAGHVGGYTESSEESNPIDFEESTHENSKHHADVVEYEEDTNPGGGQVTTESNLVEFDEESTKGIVTGAVSDHTTVEDTKEYTTESNLIELVDELPEEHGQAQGPVEEITKNNHHISHSGLGTENGHGNYDVIEEIEENSHVDIKSELGYEGGQNSGNQSFEEDTEEDKPKYEQGGNIVDIDFDSVPQIHGQNKGNQSFEEDTEKDKPKYEHGGNIIDIDFDSVPHIHGFNKHTEIIEEDTNKDKPSYQFGGHNSVDFEEDTLPKVSGQNEGQQTIEEDTTPPIVPPTPPTPEVPSEPETPTPPTPEVPSEPETPTPPTPEVPSEPETPTPPTPEVPAEPGKPVPPAKEEPKKPSKPVEQGKVVTPVIEINEKVKAVAPTKKPQSKKSELPETGGEESTNKGMLFGGLFSILGLALLRRNKKNHKA\n",
      "\n",
      "Sequence name: mRNA_5\n",
      "Protein name: mRNA_5\n",
      "Protein lenght: 401\n",
      "The starting position of ORF in RNA sequence: 311\n",
      "The ending position of ORF in RNA sequence: 1514\n",
      "Translated protein: MPLTDLKVRSLTTPGRYADGGGMYLQITASGGRYWRMKYRHGGKEKLLAFGVYPLVGIKEARDKLFYAKQALSRGEDPGATKKLAKFQAEQDAANTFKAVAEDWLDHQAGRWEAITLSRTRASLEADVYPTLGSLPISSITPRQVMDVVKKIEKRGASEIATRVLQRIKSIFRYGTTHQRIERNTMADLVPSEILKPRRVQHRAALAPAQLPEFLKKLASYEGEPHTVNALHLLMLTAARPGEVRGARWDEIDLKAKLWRIPPERMKMRSEHVVPLSSEAVAAIEAMRPLSGNRDLVFPSPFYPSKPLSENTFNSAMARMGYKYLATAHGFRALFSTVANESGWSADAIERQLAHAERNQVRAAYDRGVRLPERVKMMNWWAQYLAERKAGKAISSPEVQE\n",
      "\n",
      "Sequence name: mRNA_6\n",
      "Protein name: mRNA_6\n",
      "Protein lenght: 580\n",
      "The starting position of ORF in RNA sequence: 240\n",
      "The ending position of ORF in RNA sequence: 1980\n",
      "Translated protein: MLTACISFGVAMTTNTHFRGEELKKVWLNRYPADVPTEINPDRYQSLVDMFEQSVGALRRSTCVCEYGGGNDLRKLEERSRAFAAYLQQGLGLKKGDRVALMMPNLLQYPVALFGILRAGMIVVNVNPLYTPRELEHQLNDSGASAIVIVSNFAHTLEKVVDKTAVQHVILTRMGDQLSTAKGTVVNFVVKYIKRLVPKYHLPDAISFRSALHNGYRMQYVKPELVPEDLAFLQYTGGTTGVAKGAMLTHRNMLANLEQVNATYGPLLHPGKELVVTALPLYHIFALTINCLLFIELGGQNLLITNPRDIPGLVKELAKYPFTAITGVNTLFNALLNNKEFQQLDFSSLHLSAGGGMPVQQVVAERWVKLTGQYLLEGYGLTECAPLVSVNPYDIDYHSGSIGLPVPSTEAKLVDDDDNEVPPGQPGELCVKGPQVMLGYWQRPDATDEIIKNGWLHTGDIAVMDEEGFLRIVDRKKDMILVSGFNVYPTRLKMSSCSMVAYRKSRLLAYLPGSSGEGVKIFVVKKDPSLTEESLVTFCRRQLTGYKVPKLVEFRDELPKSNVGKILRRELRDEARQSGQ\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "D:\\Programs\\Anaconda3\\lib\\site-packages\\Bio\\Seq.py:2715: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning)\n"
     ]
    }
   ],
   "source": [
    "## ORF OUT OF THE RNA SEQUENCES\n",
    "\n",
    "# Import biopython and python modules\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import IUPAC\n",
    "import re\n",
    "from Bio import SeqIO\n",
    "\n",
    "from Bio import Seq\n",
    "\n",
    "# Save as GenBank file\n",
    "output_file = open('1-proteins.gb', 'w')\n",
    "\n",
    "# File reader as fasta\n",
    "for seq_record in SeqIO.parse(\"mRNA.fasta\",\"fasta\"):\n",
    "    # Prints out the sequence IDS\n",
    "    idSeq = seq_record.id\n",
    "    print(\"Sequence name:\",idSeq)\n",
    "    print(\"Protein name:\",idSeq)\n",
    "    \n",
    "    # Gets the sequence one at a time\n",
    "    s = seq_record.seq\n",
    "    \n",
    "    # Have to be a string not a sequence object. Otherwise it would give an error\n",
    "    madeintoString = str(s)                    \n",
    "    \n",
    "    # Start of the RNA ORF sequence\n",
    "    startCodon = re.compile('AUG')\n",
    "    \n",
    "    # Sequence string\n",
    "    nuc = madeintoString\n",
    "    \n",
    "    # Needed for comparison\n",
    "    longest = (0,)\n",
    "    \n",
    "    # ORF finder\n",
    "    # Used code by iayork --> https://stackoverflow.com/questions/31757876/python-find-longest-orf-in-dna-sequence\n",
    "    # Modified to work for RNA and modified even more to ACTUALLY work, because without the modification the original script ...\n",
    "    # ... does not work.\n",
    "    \n",
    "    # Uses regex to find the starting pattern in the sequence\n",
    "    for m in re.finditer(startCodon, nuc):\n",
    "        # Gives the longest translated protein out of every protein translated\n",
    "        if len(Seq.Seq(nuc)[m.start():].translate(to_stop=True)) > longest[0]:\n",
    "            # Translated protein\n",
    "            pro = Seq.Seq(nuc)[m.start():].translate(to_stop=True)\n",
    "            # This gives the lenght of the translated protein, that starting index of the ORF (AUG codon), \n",
    "            # the ending index of the ORF (without the position of the stopcodon included) and the protein itself\n",
    "            longest = (len(pro), \n",
    "                       m.start(), \n",
    "                       len((nuc[m.start():m.start()+len(pro)*3]))+m.start(),\n",
    "                       str(pro))\n",
    "     \n",
    "    # Protein lenght\n",
    "    lenght = longest[0]\n",
    "    print(\"Protein lenght:\",lenght)\n",
    "    \n",
    "     # Start of ORF\n",
    "    startORF = longest[1]\n",
    "    print(\"The starting position of ORF in RNA sequence:\",startORF)\n",
    "    \n",
    "    # End of ORF\n",
    "    endORF = longest[2]\n",
    "    print(\"The ending position of ORF in RNA sequence:\",endORF)\n",
    "    \n",
    "    # Protein sequence\n",
    "    prot = longest[3]\n",
    "    print(\"Translated protein:\",prot)\n",
    "    print()\n",
    "   \n",
    "    # Below is writing the file in genbank format\n",
    "    # Create a sequence\n",
    "    sequence_object = Seq.Seq(prot, IUPAC.extended_protein)\n",
    " \n",
    "    # Create a record\n",
    "    record = SeqRecord(sequence_object,\n",
    "                            id = idSeq, # random accession number which is mRNA in our case\n",
    "                            name = idSeq)\n",
    "    # Write to file\n",
    "    SeqIO.write(record, output_file, 'gb')\n",
    "\n",
    "# Close the file\n",
    "output_file.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b><u>Report/Disscussion</u></b><p>\n",
    "\n",
    "After the code has completed its run, we can see that the translated ORF of every mRNA have very different protein lenghts. mRNA_4 being the longest protein and mRNA_3 being the shortest of them all.<br> \n",
    "Because the original sequence is a RNA sequence, there is no need to find the reverse complement for the ORF because RNA is not double helical.<br>\n",
    "\n",
    "The actual approach was that first the code must read every RNA sequence from a FASTA file. The FASTA file has serveral mRNA sequences. There we can get the individual mRNA sequence and its ID. To translate the protein, its open reading frame (ORF) is needed, which in RNA's case starts with AUG codon. The actual ORF starts with the AUG and ends with its end codon, but in actual translation the end codon is not included. So we must skim trough every individual sequence and find its longest ORF. The approach is a loop. I goes through the individual mRNA sequence, looks for the AUG codon and a stop codon (from translate table, which is provided by BioPython, but since we are using the 1st one which is a starndar codon table we don't have to specify it, BioPython assumes we use the first 1 by default). After finding every ORF possible we look for the longest one and then translate it to get a protein.<br>\n",
    "Then all of the translated proteins are written into a genbank file by using BioPython provided methods. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. Similarity search\n",
    "\n",
    "Finds similar proteins from the NCBI Protein database.\n",
    "\n",
    "a) For each target protein sequence in the task 1, runs BLAST to obtain similar sequences. Saves the BLAST results to the file `2-blast.xml`.\n",
    "\n",
    "b) Selects the hits containing a high-scoring pair (HSP) with the following properties:\n",
    "- has E-value $\\leq 10^{-6}$\n",
    "- spans $\\geq 75\\%$ of the target sequence\n",
    "- percent identity $\\geq 30\\%$\n",
    "\n",
    "c) Updates the records of the target protein sequences by adding the database ids of the selected hits. Uses a sequence feature with the type `BLAST` and the qualifier key `db_xref` containing a list of ids. Saves the updated records to the file `2-proteins.gb`.\n",
    "\n",
    "d) Summarises discoveries by reporting the ids of the selected hits for each target protein sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "D:\\Programs\\Anaconda3\\lib\\site-packages\\ipykernel_launcher.py:9: DeprecationWarning: 'U' mode is deprecated\n",
      "  if __name__ == '__main__':\n"
     ]
    }
   ],
   "source": [
    "## GENBANK FILE CONVERTION INTO FASTA FILE\n",
    "\n",
    "# Import biopython and python modules\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Turns genbank file into Fasta\n",
    "# The code snipit is taken from biopython.org:\n",
    "# link --> https://biopython.org/wiki/Converting_sequence_files\n",
    "with open(\"1-proteins.gb\", \"rU\") as input_handle:\n",
    "    with open(\"1-proteins.fasta\", \"w\") as output_handle:\n",
    "        sequences = SeqIO.parse(input_handle, \"genbank\")\n",
    "        count = SeqIO.write(sequences, output_handle, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## BLAST RUN \n",
    "\n",
    "# Import biopython and python modules\n",
    "from Bio.Blast import NCBIWWW\n",
    "\n",
    "## Opens the fasta file and uses it to protein BLAST the sequences in it\n",
    "# For every proten sequence\n",
    "fastaFile = open(\"1-proteins.fasta\",\"r\")\n",
    "# Main line of code where the pBLAST happens\n",
    "result_handle = NCBIWWW.qblast(\"blastp\", \"swissprot\", fastaFile.read())\n",
    "fastaFile.close()\n",
    "\n",
    "# Writes the protein BLAST results into an xml file\n",
    "with open(\"2-blast.xml\", \"w\") as file:\n",
    "    file.write(result_handle.read())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Program: blastp (2.10.0+)\n",
      "  Query: mRNA_1 (616)\n",
      " Target: swissprot\n",
      "   Hits: ----  -----  ----------------------------------------------------------\n",
      "            #  # HSP  ID + description\n",
      "         ----  -----  ----------------------------------------------------------\n",
      "            0      1  sp|P52326.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "            1      1  sp|P26480.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "            2      1  sp|P52327.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "            3      1  sp|P0A2E3.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "            4      1  sp|P00579.2|  RecName: Full=RNA polymerase sigma factor...\n",
      "            5      1  sp|O24744.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "            6      1  sp|P32001.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "            7      1  sp|P57163.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "            8      1  sp|Q89B10.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "            9      1  sp|Q83BB6.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           10      1  sp|Q9PDM9.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           11      1  sp|Q87DT7.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           12      1  sp|Q8PG33.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           13      1  sp|P43766.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           14      1  sp|Q8P4H2.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           15      1  sp|P52325.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           16      1  sp|P33452.2|  RecName: Full=RNA polymerase sigma factor...\n",
      "           17      1  sp|Q59753.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           18      1  sp|Q2K619.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           19      1  sp|P52324.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           20      1  sp|D5AQI9.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           21      1  sp|P17531.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           22      1  sp|P0CZ15.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           23      1  sp|Q1RKH7.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           24      1  sp|Q92FZ8.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           25      1  sp|Q4UJT1.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           26      1  sp|P33451.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           27      1  sp|Q68VQ5.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           28      1  sp|P43344.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           29      2  sp|P52329.2|  RecName: Full=RNA polymerase sigma factor...\n",
      "           ~~~\n",
      "           47      2  sp|P33656.1|  RecName: Full=RNA polymerase sigma factor...\n",
      "           48      2  sp|B1VXR4.1|  RecName: Full=RNA polymerase principal si...\n",
      "           49      2  sp|P18183.2|  RecName: Full=RNA polymerase principal si...\n",
      "Program: blastp (2.10.0+)\n",
      "  Query: mRNA_2 (372)\n",
      " Target: swissprot\n",
      "   Hits: ----  -----  ----------------------------------------------------------\n",
      "            #  # HSP  ID + description\n",
      "         ----  -----  ----------------------------------------------------------\n",
      "            0      1  sp|P9WK86.1|  RecName: Full=Carboxylesterase NlhH [Myco...\n",
      "            1      1  sp|P24484.1|  RecName: Full=Lipase 2; AltName: Full=Tri...\n",
      "            2      1  sp|Q5UQ83.1|  RecName: Full=Putative alpha/beta hydrola...\n",
      "            3      1  sp|B5BLW5.1|  RecName: Full=Arylesterase; Short=A-ester...\n",
      "            4      1  sp|Q9US38.1|  RecName: Full=AB hydrolase superfamily pr...\n",
      "            5      1  sp|Q0ZPV7.1|  RecName: Full=Carboxylesterase 1; Short=A...\n",
      "            6      1  sp|R4X5P0.1|  RecName: Full=Tuliposide A-converting enz...\n",
      "            7      1  sp|Q9LYC1.1|  RecName: Full=Gibberellin receptor GID1B;...\n",
      "            8      1  sp|R4X244.1|  RecName: Full=Tuliposide A-converting enz...\n",
      "            9      1  sp|R4X4V6.1|  RecName: Full=Tuliposide A-converting enz...\n",
      "           10      1  sp|R4X247.1|  RecName: Full=Probable tuliposide A-conve...\n",
      "           11      1  sp|Q9HDX3.1|  RecName: Full=AB hydrolase superfamily pr...\n",
      "           12      1  sp|Q940G6.1|  RecName: Full=Gibberellin receptor GID1C;...\n",
      "           13      1  sp|Q9MAA7.1|  RecName: Full=Gibberellin receptor GID1A;...\n",
      "           14      1  sp|P14326.2|  RecName: Full=Vegetative-specific protein...\n",
      "           15      1  sp|M2YJ38.1|  RecName: Full=Versiconal hemiacetal aceta...\n",
      "           16      1  sp|P18773.2|  RecName: Full=Esterase [Acinetobacter lwo...\n",
      "           17      1  sp|I4DST8.1|  RecName: Full=Tuliposide A-converting enz...\n",
      "           18      1  sp|I4DST9.1|  RecName: Full=Tuliposide A-converting enz...\n",
      "           19      2  sp|P15304.3|  RecName: Full=Hormone-sensitive lipase; S...\n",
      "           20      1  sp|I3PLR2.1|  RecName: Full=3-O-acetylpapaveroxine carb...\n",
      "           21      1  sp|P22760.5|  RecName: Full=Arylacetamide deacetylase [...\n",
      "           22      1  sp|Q8BLF1.1|  RecName: Full=Neutral cholesterol ester h...\n",
      "           23      2  sp|Q05469.4|  RecName: Full=Hormone-sensitive lipase; S...\n",
      "           24      1  sp|Q1JQE6.2|  RecName: Full=Neutral cholesterol ester h...\n",
      "           25      1  sp|O64640.1|  RecName: Full=Probable carboxylesterase 8...\n",
      "           26      1  sp|Q9SX25.1|  RecName: Full=Probable carboxylesterase 6...\n",
      "           27      2  sp|P54310.2|  RecName: Full=Hormone-sensitive lipase; S...\n",
      "           28      1  sp|Q5VUY0.5|  RecName: Full=Arylacetamide deacetylase-l...\n",
      "           29      1  sp|A0A0A1EQ07.1|  RecName: Full=3-O-acetylpapaveroxine ...\n",
      "           ~~~\n",
      "           47      1  sp|Q8Z8T1.1|  RecName: Full=Acetyl esterase [Salmonella...\n",
      "           48      1  sp|B4TMG8.1|  RecName: Full=Acetyl esterase [Salmonella...\n",
      "           49      1  sp|Q9SMM9.1|  RecName: Full=Probable carboxylesterase 1...\n",
      "Program: blastp (2.10.0+)\n",
      "  Query: mRNA_3 (348)\n",
      " Target: swissprot\n",
      "   Hits: ----  -----  ----------------------------------------------------------\n",
      "            #  # HSP  ID + description\n",
      "         ----  -----  ----------------------------------------------------------\n",
      "            0      1  sp|Q9RR29.1|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "            1      1  sp|D4GU70.1|  RecName: Full=Low-salt glycan biosynthesi...\n",
      "            2      1  sp|P08075.1|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "            3      1  sp|A6VG23.1|  RecName: Full=Bifunctional protein GlmU; ...\n",
      "            4      1  sp|A6UP85.1|  RecName: Full=Bifunctional protein GlmU; ...\n",
      "            5      1  sp|A4FX98.1|  RecName: Full=Bifunctional protein GlmU; ...\n",
      "            6      1  sp|Q58501.1|  RecName: Full=Bifunctional protein GlmU; ...\n",
      "            7      1  sp|Q6LYB5.1|  RecName: Full=Bifunctional protein GlmU; ...\n",
      "            8      1  sp|A6UUQ4.1|  RecName: Full=Bifunctional protein GlmU; ...\n",
      "            9      1  sp|D4GYH1.1|  RecName: Full=UTP--glucose-1-phosphate ur...\n",
      "           10      1  sp|Q975F9.1|  RecName: Full=Bifunctional sugar-1-phosph...\n",
      "           11      1  sp|P39629.1|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "           12      1  sp|P95778.1|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "           13      1  sp|P37779.2|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "           14      1  sp|P55255.2|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "           15      1  sp|P57040.1|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "           16      1  sp|P37744.2|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "           17      1  sp|P61887.1|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "           18      1  sp|P55253.1|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "           19      1  sp|P55464.1|  RecName: Full=Probable glucose-1-phosphat...\n",
      "           20      1  sp|P55254.1|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "           21      1  sp|P26393.1|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "           22      1  sp|A0QPF9.1|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "           23      1  sp|P9WH12.1|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "           24      1  sp|P37762.1|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "           25      1  sp|Q9L9E6.1|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "           26      1  sp|B0RVK9.1|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "           27      1  sp|P55257.1|  RecName: Full=Glucose-1-phosphate thymidy...\n",
      "           28      1  sp|Q9Y725.1|  RecName: Full=Mannose-1-phosphate guanylt...\n",
      "           29      1  sp|Q05852.1|  RecName: Full=UTP--glucose-1-phosphate ur...\n",
      "           ~~~\n",
      "           47      1  sp|A2VD83.1|  RecName: Full=Mannose-1-phosphate guanylt...\n",
      "           48      1  sp|Q295Y7.1|  RecName: Full=Mannose-1-phosphate guanylt...\n",
      "           49      1  sp|Q4U3E8.1|  RecName: Full=Mannose-1-phosphate guanylt...\n",
      "Program: blastp (2.10.0+)\n",
      "  Query: mRNA_4 (995)\n",
      " Target: swissprot\n",
      "   Hits: ----  -----  ----------------------------------------------------------\n",
      "            #  # HSP  ID + description\n",
      "         ----  -----  ----------------------------------------------------------\n",
      "            0      1  sp|P14738.1|  RecName: Full=Fibronectin-binding protein...\n",
      "            1      1  sp|Q2FE03.1|  RecName: Full=Fibronectin-binding protein...\n",
      "            2      1  sp|Q6G6H3.1|  RecName: Full=Fibronectin-binding protein...\n",
      "            3      1  sp|A7X6I5.1|  RecName: Full=Fibronectin-binding protein...\n",
      "            4      1  sp|Q6GDU5.1|  RecName: Full=Fibronectin-binding protein...\n",
      "            5      1  sp|Q2YW62.1|  RecName: Full=Fibronectin-binding protein...\n",
      "            6      1  sp|A0A0H2XKG3.1|  RecName: Full=Fibronectin-binding pro...\n",
      "            7      1  sp|Q6GIK4.1|  RecName: Full=Clumping factor A; AltName:...\n",
      "            8      1  sp|Q932C5.1|  RecName: Full=Clumping factor A; AltName:...\n",
      "            9      1  sp|Q99VJ4.1|  RecName: Full=Clumping factor A; AltName:...\n",
      "           10      2  sp|Q6GB45.1|  RecName: Full=Clumping factor A; AltName:...\n",
      "           11      2  sp|Q8NXJ1.1|  RecName: Full=Clumping factor A; AltName:...\n",
      "           12      2  sp|Q5HHM8.1|  RecName: Full=Clumping factor A; AltName:...\n",
      "           13      2  sp|Q2G015.1|  RecName: Full=Clumping factor A; AltName:...\n",
      "           14      2  sp|Q53653.1|  RecName: Full=Clumping factor A; AltName:...\n",
      "           15      2  sp|Q6GDH2.1|  RecName: Full=Clumping factor B; AltName:...\n",
      "           16      1  sp|Q70XJ9.1|  RecName: Full=Levansucrase; Flags: Precur...\n",
      "           17      2  sp|Q6G644.1|  RecName: Full=Clumping factor B; AltName:...\n",
      "           18      2  sp|Q8NUL0.1|  RecName: Full=Clumping factor B; AltName:...\n",
      "           19      2  sp|Q2FUY2.1|  RecName: Full=Clumping factor B; AltName:...\n",
      "           20      2  sp|Q7A382.1|  RecName: Full=Clumping factor B; AltName:...\n",
      "           21      2  sp|O86476.2|  RecName: Full=Clumping factor B; AltName:...\n",
      "           22      1  sp|Q8CQD9.1|  RecName: Full=Accumulation-associated pro...\n",
      "           23      2  sp|P27951.1|  RecName: Full=IgA FC receptor; AltName: F...\n",
      "           24      1  sp|Q9KI13.1|  RecName: Full=Serine-aspartate repeat-con...\n",
      "           25      2  sp|Q6GJA6.1|  RecName: Full=Bone sialoprotein-binding p...\n",
      "           26      2  sp|Q14U76.1|  RecName: Full=Bone sialoprotein-binding p...\n",
      "           27      1  sp|P16053.2|  RecName: Full=Neurofilament medium polype...\n",
      "           28      1  sp|Q931E9.2|  RecName: Full=Putative surface protein SA...\n",
      "           29      2  sp|P16952.2|  RecName: Full=Agglutinin receptor; AltNam...\n",
      "           ~~~\n",
      "           47      1  sp|Q8NUV0.1|  RecName: Full=Putative surface protein MW...\n",
      "           48      3  sp|A2ASS6.1|  RecName: Full=Titin; AltName: Full=Connec...\n",
      "           49      1  sp|Q8NXX7.1|  RecName: Full=Serine-aspartate repeat-con...\n",
      "Program: blastp (2.10.0+)\n",
      "  Query: mRNA_5 (401)\n",
      " Target: swissprot\n",
      "   Hits: ----  -----  ----------------------------------------------------------\n",
      "            #  # HSP  ID + description\n",
      "         ----  -----  ----------------------------------------------------------\n",
      "            0      1  sp|P08320.2|  RecName: Full=Integrase [Enterobacteria p...\n",
      "            1      1  sp|P39347.1|  PUTATIVE PSEUDOGENE: RecName: Full=Putati...\n",
      "            2      1  sp|P37317.1|  RecName: Full=Integrase [Shigella virus Sf6]\n",
      "            3      1  sp|P37326.1|  RecName: Full=Prophage integrase IntS; Al...\n",
      "            4      1  sp|P32053.2|  RecName: Full=Prophage integrase IntA; Al...\n",
      "            5      1  sp|P76542.2|  RecName: Full=Prophage integrase IntZ; Al...\n",
      "            6      1  sp|P06155.1|  RecName: Full=Integrase [Enterobacteria p...\n",
      "            7      1  sp|P42540.1|  RecName: Full=Probable integrase/recombin...\n",
      "            8      1  sp|P71298.1|  RecName: Full=Prophage integrase IntF; Al...\n",
      "            9      1  sp|Q9Z9F7.1|  RecName: Full=Tyrosine recombinase XerC [...\n",
      "           10      1  sp|P0A053.1|  RecName: Full=Transposase B from transpos...\n",
      "           11      1  sp|P37375.1|  RecName: Full=Transposase B from transpos...\n",
      "           12      1  sp|B2IA18.1|  RecName: Full=Tyrosine recombinase XerC [...\n",
      "           13      1  sp|P76168.3|  RecName: Full=Putative defective protein ...\n",
      "           14      1  sp|P0A050.1|  RecName: Full=Transposase A from transpos...\n",
      "           15      1  sp|P0C9V1.1|  RecName: Full=Protein MGF 505-10R [Africa...\n",
      "           16      1  sp|P0C9V2.1|  RecName: Full=Protein MGF 505-10R [Africa...\n",
      "           17      1  sp|Q89869.1|  RecName: Full=Protein MGF 505-10R [Africa...\n",
      "           18      1  sp|Q9PD96.1|  RecName: Full=Tyrosine recombinase XerC [...\n",
      "           19      1  sp|P46352.3|  RecName: Full=Tyrosine recombinase XerD [...\n",
      "           20      1  sp|P16470.1|  RecName: Full=Shufflon-specific DNA recom...\n",
      "           21      1  sp|P10487.1|  RecName: Full=Shufflon-specific DNA recom...\n",
      "           22      1  sp|B0RNK3.1|  RecName: Full=Tyrosine recombinase XerC [...\n",
      "           23      1  sp|A7NFG3.1|  RecName: Full=Tyrosine recombinase XerC [...\n",
      "           24      1  sp|B2UPN5.1|  RecName: Full=2-succinyl-5-enolpyruvyl-6-...\n",
      "           25      1  sp|Q9M358.1|  RecName: Full=UV-stimulated scaffold prot...\n",
      "           26      1  sp|P45198.1|  RecName: Full=Putative integrase/recombin...\n",
      "           27      1  sp|P25641.2|  RecName: Full=Putative lipase ATG15; AltN...\n",
      "           28      1  sp|A6ZTP2.1|  RecName: Full=Putative lipase ATG15; AltN...\n",
      "           29      1  sp|Q67MV8.1|  RecName: Full=Alanine--tRNA ligase; AltNa...\n",
      "           30      1  sp|A6UT10.1|  RecName: Full=Cobalt-precorrin-5B C(1)-me...\n",
      "           31      1  sp|Q8IZH2.1|  RecName: Full=5'-3' exoribonuclease 1; Al...\n",
      "           32      1  sp|Q9KCP0.1|  RecName: Full=Tyrosine recombinase XerD [...\n",
      "Program: blastp (2.10.0+)\n",
      "  Query: mRNA_6 (580)\n",
      " Target: swissprot\n",
      "   Hits: ----  -----  ----------------------------------------------------------\n",
      "            #  # HSP  ID + description\n",
      "         ----  -----  ----------------------------------------------------------\n",
      "            0      1  sp|P69451.1|  RecName: Full=Long-chain-fatty-acid--CoA ...\n",
      "            1      1  sp|Q8XDR6.1|  RecName: Full=Long-chain-fatty-acid--CoA ...\n",
      "            2      1  sp|P63521.1|  RecName: Full=Long-chain-fatty-acid--CoA ...\n",
      "            3      1  sp|Q8ZES9.1|  RecName: Full=Long-chain-fatty-acid--CoA ...\n",
      "            4      1  sp|P46450.1|  RecName: Full=Long-chain-fatty-acid--CoA ...\n",
      "            5      1  sp|P94547.1|  RecName: Full=Long-chain-fatty-acid--CoA ...\n",
      "            6      1  sp|O07610.2|  RecName: Full=Long-chain-fatty-acid--CoA ...\n",
      "            7      1  sp|Q0DV32.2|  RecName: Full=4-coumarate--CoA ligase-lik...\n",
      "            8      1  sp|O24145.1|  RecName: Full=4-coumarate--CoA ligase 1; ...\n",
      "            9      1  sp|P41636.1|  RecName: Full=4-coumarate--CoA ligase; Sh...\n",
      "           10      1  sp|Q499N5.1|  RecName: Full=Medium-chain acyl-CoA ligas...\n",
      "           11      1  sp|Q9S725.2|  RecName: Full=4-coumarate--CoA ligase 2; ...\n",
      "           12      1  sp|Q42982.2|  RecName: Full=Probable 4-coumarate--CoA l...\n",
      "           13      1  sp|Q9M0X9.1|  RecName: Full=4-coumarate--CoA ligase-lik...\n",
      "           14      1  sp|Q6ETN3.1|  RecName: Full=Probable 4-coumarate--CoA l...\n",
      "           15      1  sp|O31826.1|  RecName: Full=Putative acyl-CoA synthetas...\n",
      "           16      1  sp|Q7WSH3.1|  RecName: Full=3-[(3aS,4S,7aS)-7a-methyl-1...\n",
      "           17      1  sp|Q6ZAC1.1|  RecName: Full=Probable 4-coumarate--CoA l...\n",
      "           18      1  sp|Q17QJ1.1|  RecName: Full=Medium-chain acyl-CoA ligas...\n",
      "           19      1  sp|P31684.1|  RecName: Full=4-coumarate--CoA ligase 1; ...\n",
      "           20      1  sp|P31685.1|  RecName: Full=4-coumarate--CoA ligase 2; ...\n",
      "           21      1  sp|Q0S7V5.1|  RecName: Full=3-[(3aS,4S,7aS)-7a-methyl-1...\n",
      "           22      1  sp|Q67W82.1|  RecName: Full=Probable 4-coumarate--CoA l...\n",
      "           23      1  sp|Q8VCW8.1|  RecName: Full=Medium-chain acyl-CoA ligas...\n",
      "           24      1  sp|Q84P24.2|  RecName: Full=4-coumarate--CoA ligase-lik...\n",
      "           25      1  sp|P96575.2|  RecName: Full=Putative acyl--CoA ligase Y...\n",
      "           26      1  sp|Q7XXL2.2|  RecName: Full=4-coumarate--CoA ligase-lik...\n",
      "           27      1  sp|Q4R4Z9.1|  RecName: Full=Medium-chain acyl-CoA ligas...\n",
      "           28      1  sp|Q84P21.2|  RecName: Full=4-coumarate--CoA ligase-lik...\n",
      "           29      1  sp|Q84P25.2|  RecName: Full=4-coumarate--CoA ligase-lik...\n",
      "           ~~~\n",
      "           47      1  sp|Q0P4F7.1|  RecName: Full=Medium-chain acyl-CoA ligas...\n",
      "           48      1  sp|A9VM74.1|  RecName: Full=2-succinylbenzoate--CoA lig...\n",
      "           49      1  sp|Q816I1.1|  RecName: Full=2-succinylbenzoate--CoA lig...\n"
     ]
    }
   ],
   "source": [
    "## BLAST HIT READING\n",
    "\n",
    "# Import bipython modules\n",
    "import Bio.SearchIO as BSIO\n",
    "\n",
    "# Reads all of the pBLAST results that are in the xml file\n",
    "with open('2-blast.xml') as f:\n",
    "    for result in BSIO.parse(f, 'blast-xml'):\n",
    "        print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein name: mRNA_1\n",
      "Blast hit ID's: ['P52326', 'P26480', 'P52327', 'P0A2E3', 'P00579', 'O24744', 'P32001', 'P57163', 'Q89B10', 'Q83BB6', 'Q9PDM9', 'Q87DT7', 'Q8PG33', 'P43766', 'Q8P4H2', 'P52325', 'P33452', 'Q59753', 'Q2K619', 'P52324', 'D5AQI9', 'P17531', 'P0CZ15', 'Q1RKH7', 'Q92FZ8', 'Q4UJT1', 'P33451', 'Q68VQ5']\n",
      "\n",
      "Protein name: mRNA_2\n",
      "Blast hit ID's: []\n",
      "\n",
      "Protein name: mRNA_3\n",
      "Blast hit ID's: ['Q9RR29', 'D4GU70', 'P08075', 'A6VG23', 'A6UP85', 'A4FX98', 'Q58501', 'Q6LYB5', 'A6UUQ4', 'Q975F9', 'Q9Y725', 'Q6FRY2', 'P0C5I2', 'O93827', 'Q8BTZ7', 'Q9Y5P6', 'Q752H4', 'P41940', 'Q6BN12', 'O74624', 'Q2YDJ9', 'A2VD83', 'Q4U3E8']\n",
      "\n",
      "Protein name: mRNA_4\n",
      "Blast hit ID's: ['P14738', 'Q2FE03', 'Q6G6H3', 'A7X6I5', 'Q6GDU5', 'Q2YW62', 'A0A0H2XKG3']\n",
      "\n",
      "Protein name: mRNA_5\n",
      "Blast hit ID's: ['P08320', 'P39347', 'P37317', 'P37326', 'P32053', 'P76542']\n",
      "\n",
      "Protein name: mRNA_6\n",
      "Blast hit ID's: ['P69451', 'Q8XDR6', 'P63521', 'Q8ZES9', 'P46450', 'P94547', 'O07610', 'Q0DV32', 'P41636', 'Q42982', 'Q54P79', 'Q54P77', 'P14912', 'P14913', 'O24146', 'Q54P78']\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## FILTERING BLAST HITS AND FILE WRITING\n",
    "\n",
    "# Import biopython and python modules\n",
    "from Bio import SeqIO\n",
    "import Bio.SearchIO as BSIO\n",
    "from Bio import SeqFeature\n",
    "from Bio.SeqFeature import SeqFeature\n",
    "from Bio import SeqFeature as sf\n",
    "import Bio.SeqFeature as BSF\n",
    "\n",
    "# Visas tas sarasas tu proteinu\n",
    "# All of the existing genbank mRNAs with their protein\n",
    "proteins = list(SeqIO.parse(\"1-proteins.gb\",\"gb\"))\n",
    "#print(proteins)\n",
    "\n",
    "# Function for e-value that is lest or equel to 10e-6\n",
    "def evalue(hsp):\n",
    "    return hsp.evalue <= 0.000001\n",
    "     \n",
    "# Function for span that is more or equel to 75 percent of the target sequence\n",
    "def span(hsp):\n",
    "    return (hsp.query_span / seq_len) * 100 >= 75\n",
    "\n",
    "# Function for percent identity that is more or equal to 30\n",
    "def identity(hsp):\n",
    "    return (hsp.ident_num / hsp.aln_span) * 100 >= 30\n",
    "\n",
    "# Main code that reads the file \n",
    "with open('2-blast.xml') as file:\n",
    "    for result in BSIO.parse(file, 'blast-xml'):\n",
    "        # Filteres all of the results that have that evalue()\n",
    "        filtered_result = result.hsp_filter(evalue)\n",
    "        # Sequnce lenghts with that evalue\n",
    "        seq_len = filtered_result.seq_len\n",
    "        # Filteres out the hits that have that evalue() filtered before but this time looks at the span()\n",
    "        results_forvalueNspan = filtered_result.hsp_filter(span)\n",
    "        # Filteres out the hits that have that evalue() and span() like filtered before but this time looks at the identity()\n",
    "        results_forAll = results_forvalueNspan.hsp_filter(identity)\n",
    "        #for hit in results_forAll:\n",
    "            #print(hit)\n",
    "        \n",
    "        # every mRNA result ID name and so on from BLAST\n",
    "        rnaName = result.id\n",
    "        print(\"Protein name:\",rnaName)\n",
    "        \n",
    "        # Prints out all of the results that were filtered\n",
    "        db_xrefs_seq_ids = []\n",
    "        \n",
    "        # Gets all of the hit IDs without the sp|| around it. Thats why we use accession\n",
    "        # Appends into a list\n",
    "        for hit in results_forAll:\n",
    "            idd = hit.accession\n",
    "            db_xrefs_seq_ids.append(idd)\n",
    "        print(\"Blast hit ID's:\",db_xrefs_seq_ids)\n",
    "        print()\n",
    "            \n",
    "        # Check if we found it or not\n",
    "        protein = None\n",
    "        \n",
    "        # jei liste tas mano s turi ta name = \"\"\n",
    "        # So for every protein or that mRNA1 in all of the proteins or alll of the mRNAs in the file we look \n",
    "        # if the name  of the protein in genbank file is the same as the name in the BLAST results of hits\n",
    "        # proteins is our genbank file\n",
    "        for p in proteins:\n",
    "            if p.name == rnaName:\n",
    "                protein = p\n",
    "                break\n",
    "        #print(protein)\n",
    "            \n",
    "        # Creation of BLAST feature for our protein sequences'\n",
    "        # Len vieno protein o ne visu proteins yra tas 1..616 salia BLAST\n",
    "        feature = BSF.SeqFeature(BSF.FeatureLocation(0,len(protein)), type='BLAST')\n",
    "        # We add a qualifier 'db_xref' to the hit ids (accenssions) and add it to BLAST feature\n",
    "        feature.qualifiers['db_xref'] = db_xrefs_seq_ids\n",
    "        \n",
    "        # Then for every protein (not the whole stack of proteins but just every individual protein/mRNA)in the genbank file \n",
    "        # we add those features that have a name BLAST on the left collum\n",
    "        # and we add all of the feature items, I mean all of those ids with the qualifier\n",
    "        protein.features.append(feature)\n",
    "        #print(features)\n",
    "\n",
    "# Write to genbank file\n",
    "SeqIO.write(proteins, \"2-proteins.gb\",\"gb\")\n",
    "      "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b><u>Report/Disscussion</u></b><p>\n",
    "As we can see after the run that only the mRNA2 protein does not have BLAST IDs based on our criteria.<p>\n",
    "The base idea of the code was to find BLAST hits that are based on 3 properties.<br>\n",
    "First I converted the GenBank file into FASTA file to get the sequences we need. Then for every mRNA protein sequence I ran protein BLAST in swissprot library. Then I checked the given blastp results. To get the hits based on our criteria I made functions for every property like: evalue for e-value, span for aligned sequence span and identity for sequence similarity. The functions make the code more understandable. I used those functions to filter out the results that are stored in the blast xml file.<br>\n",
    "Then I took out just the IDs from our filtered BLAST results. Then I checked if our filtered BLAST results have the same query name like the ones in our genbank file. Basically for example if mRNA_1 in BLAST results equel to mRNA1 in our genbank file.<br>\n",
    "Later I created a new genbank file sequence Feature called BLAST, there I add qualifiers that are our BLAST IDs. Then append those features to every protein in genbank file. And then parse it into a new file: take the information form our 1st genbank file and make a new one (2nd file)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3. Protein function prediction\n",
    "\n",
    "Infering of the putative functions by similarity.\n",
    "\n",
    "a) Downloads the UniProt entries of the selected hits in the task 2. Saves the entries to the files `XXXXXX.xml` where `XXXXXX` is the UniProt id.\n",
    "\n",
    "b) Separately for each target protein sequence in the task 1, finds those Gene Ontology (GO) terms which occur `in all` of the matched UniProt entries.\n",
    "\n",
    "c) Updates the records of the target protein sequences by adding the found GO terms. Uses a sequence feature with the type GO and the qualifier key `db_xref` containing a list of ids. Also adds the qualifier `evidence` with value `inferred by similarity` to the feature. Saves the updated records to the file `3-proteins.gb`.\n",
    "\n",
    "d) Summarises discoveries by reporting the putative GO terms for each target protein sequence. Prints both the GO term ids and their human-friendly names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['P52326', 'P26480', 'P52327', 'P0A2E3', 'P00579', 'O24744', 'P32001', 'P57163', 'Q89B10', 'Q83BB6', 'Q9PDM9', 'Q87DT7', 'Q8PG33', 'P43766', 'Q8P4H2', 'P52325', 'P33452', 'Q59753', 'Q2K619', 'P52324', 'D5AQI9', 'P17531', 'P0CZ15', 'Q1RKH7', 'Q92FZ8', 'Q4UJT1', 'P33451', 'Q68VQ5']\n",
      "P52326\n",
      "https://www.uniprot.org/uniprot/P52326.xml\n",
      "P26480\n",
      "https://www.uniprot.org/uniprot/P26480.xml\n",
      "P52327\n",
      "https://www.uniprot.org/uniprot/P52327.xml\n",
      "P0A2E3\n",
      "https://www.uniprot.org/uniprot/P0A2E3.xml\n",
      "P00579\n",
      "https://www.uniprot.org/uniprot/P00579.xml\n",
      "O24744\n",
      "https://www.uniprot.org/uniprot/O24744.xml\n",
      "P32001\n",
      "https://www.uniprot.org/uniprot/P32001.xml\n",
      "P57163\n",
      "https://www.uniprot.org/uniprot/P57163.xml\n",
      "Q89B10\n",
      "https://www.uniprot.org/uniprot/Q89B10.xml\n",
      "Q83BB6\n",
      "https://www.uniprot.org/uniprot/Q83BB6.xml\n",
      "Q9PDM9\n",
      "https://www.uniprot.org/uniprot/Q9PDM9.xml\n",
      "Q87DT7\n",
      "https://www.uniprot.org/uniprot/Q87DT7.xml\n",
      "Q8PG33\n",
      "https://www.uniprot.org/uniprot/Q8PG33.xml\n",
      "P43766\n",
      "https://www.uniprot.org/uniprot/P43766.xml\n",
      "Q8P4H2\n",
      "https://www.uniprot.org/uniprot/Q8P4H2.xml\n",
      "P52325\n",
      "https://www.uniprot.org/uniprot/P52325.xml\n",
      "P33452\n",
      "https://www.uniprot.org/uniprot/P33452.xml\n",
      "Q59753\n",
      "https://www.uniprot.org/uniprot/Q59753.xml\n",
      "Q2K619\n",
      "https://www.uniprot.org/uniprot/Q2K619.xml\n",
      "P52324\n",
      "https://www.uniprot.org/uniprot/P52324.xml\n",
      "D5AQI9\n",
      "https://www.uniprot.org/uniprot/D5AQI9.xml\n",
      "P17531\n",
      "https://www.uniprot.org/uniprot/P17531.xml\n",
      "P0CZ15\n",
      "https://www.uniprot.org/uniprot/P0CZ15.xml\n",
      "Q1RKH7\n",
      "https://www.uniprot.org/uniprot/Q1RKH7.xml\n",
      "Q92FZ8\n",
      "https://www.uniprot.org/uniprot/Q92FZ8.xml\n",
      "Q4UJT1\n",
      "https://www.uniprot.org/uniprot/Q4UJT1.xml\n",
      "P33451\n",
      "https://www.uniprot.org/uniprot/P33451.xml\n",
      "Q68VQ5\n",
      "https://www.uniprot.org/uniprot/Q68VQ5.xml\n",
      "['Q9RR29', 'D4GU70', 'P08075', 'A6VG23', 'A6UP85', 'A4FX98', 'Q58501', 'Q6LYB5', 'A6UUQ4', 'Q975F9', 'Q9Y725', 'Q6FRY2', 'P0C5I2', 'O93827', 'Q8BTZ7', 'Q9Y5P6', 'Q752H4', 'P41940', 'Q6BN12', 'O74624', 'Q2YDJ9', 'A2VD83', 'Q4U3E8']\n",
      "Q9RR29\n",
      "https://www.uniprot.org/uniprot/Q9RR29.xml\n",
      "D4GU70\n",
      "https://www.uniprot.org/uniprot/D4GU70.xml\n",
      "P08075\n",
      "https://www.uniprot.org/uniprot/P08075.xml\n",
      "A6VG23\n",
      "https://www.uniprot.org/uniprot/A6VG23.xml\n",
      "A6UP85\n",
      "https://www.uniprot.org/uniprot/A6UP85.xml\n",
      "A4FX98\n",
      "https://www.uniprot.org/uniprot/A4FX98.xml\n",
      "Q58501\n",
      "https://www.uniprot.org/uniprot/Q58501.xml\n",
      "Q6LYB5\n",
      "https://www.uniprot.org/uniprot/Q6LYB5.xml\n",
      "A6UUQ4\n",
      "https://www.uniprot.org/uniprot/A6UUQ4.xml\n",
      "Q975F9\n",
      "https://www.uniprot.org/uniprot/Q975F9.xml\n",
      "Q9Y725\n",
      "https://www.uniprot.org/uniprot/Q9Y725.xml\n",
      "Q6FRY2\n",
      "https://www.uniprot.org/uniprot/Q6FRY2.xml\n",
      "P0C5I2\n",
      "https://www.uniprot.org/uniprot/P0C5I2.xml\n",
      "O93827\n",
      "https://www.uniprot.org/uniprot/O93827.xml\n",
      "Q8BTZ7\n",
      "https://www.uniprot.org/uniprot/Q8BTZ7.xml\n",
      "Q9Y5P6\n",
      "https://www.uniprot.org/uniprot/Q9Y5P6.xml\n",
      "Q752H4\n",
      "https://www.uniprot.org/uniprot/Q752H4.xml\n",
      "P41940\n",
      "https://www.uniprot.org/uniprot/P41940.xml\n",
      "Q6BN12\n",
      "https://www.uniprot.org/uniprot/Q6BN12.xml\n",
      "O74624\n",
      "https://www.uniprot.org/uniprot/O74624.xml\n",
      "Q2YDJ9\n",
      "https://www.uniprot.org/uniprot/Q2YDJ9.xml\n",
      "A2VD83\n",
      "https://www.uniprot.org/uniprot/A2VD83.xml\n",
      "Q4U3E8\n",
      "https://www.uniprot.org/uniprot/Q4U3E8.xml\n",
      "['P14738', 'Q2FE03', 'Q6G6H3', 'A7X6I5', 'Q6GDU5', 'Q2YW62', 'A0A0H2XKG3']\n",
      "P14738\n",
      "https://www.uniprot.org/uniprot/P14738.xml\n",
      "Q2FE03\n",
      "https://www.uniprot.org/uniprot/Q2FE03.xml\n",
      "Q6G6H3\n",
      "https://www.uniprot.org/uniprot/Q6G6H3.xml\n",
      "A7X6I5\n",
      "https://www.uniprot.org/uniprot/A7X6I5.xml\n",
      "Q6GDU5\n",
      "https://www.uniprot.org/uniprot/Q6GDU5.xml\n",
      "Q2YW62\n",
      "https://www.uniprot.org/uniprot/Q2YW62.xml\n",
      "A0A0H2XKG3\n",
      "https://www.uniprot.org/uniprot/A0A0H2XKG3.xml\n",
      "['P08320', 'P39347', 'P37317', 'P37326', 'P32053', 'P76542']\n",
      "P08320\n",
      "https://www.uniprot.org/uniprot/P08320.xml\n",
      "P39347\n",
      "https://www.uniprot.org/uniprot/P39347.xml\n",
      "P37317\n",
      "https://www.uniprot.org/uniprot/P37317.xml\n",
      "P37326\n",
      "https://www.uniprot.org/uniprot/P37326.xml\n",
      "P32053\n",
      "https://www.uniprot.org/uniprot/P32053.xml\n",
      "P76542\n",
      "https://www.uniprot.org/uniprot/P76542.xml\n",
      "['P69451', 'Q8XDR6', 'P63521', 'Q8ZES9', 'P46450', 'P94547', 'O07610', 'Q0DV32', 'P41636', 'Q42982', 'Q54P79', 'Q54P77', 'P14912', 'P14913', 'O24146', 'Q54P78']\n",
      "P69451\n",
      "https://www.uniprot.org/uniprot/P69451.xml\n",
      "Q8XDR6\n",
      "https://www.uniprot.org/uniprot/Q8XDR6.xml\n",
      "P63521\n",
      "https://www.uniprot.org/uniprot/P63521.xml\n",
      "Q8ZES9\n",
      "https://www.uniprot.org/uniprot/Q8ZES9.xml\n",
      "P46450\n",
      "https://www.uniprot.org/uniprot/P46450.xml\n",
      "P94547\n",
      "https://www.uniprot.org/uniprot/P94547.xml\n",
      "O07610\n",
      "https://www.uniprot.org/uniprot/O07610.xml\n",
      "Q0DV32\n",
      "https://www.uniprot.org/uniprot/Q0DV32.xml\n",
      "P41636\n",
      "https://www.uniprot.org/uniprot/P41636.xml\n",
      "Q42982\n",
      "https://www.uniprot.org/uniprot/Q42982.xml\n",
      "Q54P79\n",
      "https://www.uniprot.org/uniprot/Q54P79.xml\n",
      "Q54P77\n",
      "https://www.uniprot.org/uniprot/Q54P77.xml\n",
      "P14912\n",
      "https://www.uniprot.org/uniprot/P14912.xml\n",
      "P14913\n",
      "https://www.uniprot.org/uniprot/P14913.xml\n",
      "O24146\n",
      "https://www.uniprot.org/uniprot/O24146.xml\n",
      "Q54P78\n",
      "https://www.uniprot.org/uniprot/Q54P78.xml\n"
     ]
    }
   ],
   "source": [
    "## ## READS GENBANK FILE AND SAVES BLAST IDS INTO AN XML FILE\n",
    "\n",
    "# Import biopython and python modules\n",
    "from Bio import SeqIO\n",
    "import collections as C\n",
    "import requests as R\n",
    "import itertools\n",
    "import collections\n",
    "import Bio.SeqFeature as BSF\n",
    "\n",
    "proteins2 = list(SeqIO.parse(\"2-proteins.gb\",\"gb\"))\n",
    "#print(proteins2)\n",
    "\n",
    "for protein in proteins2:\n",
    "    #print(protein)\n",
    "    # Finds all of the ID that are in genbank file 2-proteins-tryout with feature - BLAST\n",
    "    for f in protein.features:\n",
    "        if f.type == \"BLAST\":\n",
    "            if \"db_xref\" in f.qualifiers:\n",
    "                idss = f.qualifiers[\"db_xref\"]\n",
    "                print(idss)\n",
    "                \n",
    "                \n",
    "                # By the individual ID get the UniProt URLs and download the entries as XML files\n",
    "                for idd in idss:\n",
    "                    # Prints ids just to check how they look\n",
    "                    print(idd)\n",
    "\n",
    "                    # API address\n",
    "                    base = \"https://www.uniprot.org/uniprot/%s.%s\"\n",
    "\n",
    "                    # Format\n",
    "                    fmt = 'xml'\n",
    "\n",
    "                    # Adds the base url to idd and xml format by replaces the %s.%s\n",
    "                    # It would be something like this if the id (idd) is 11:\n",
    "                    # https://www.uniprot.org/uniprot/11.xml\n",
    "                    url = base%(idd, fmt)\n",
    "\n",
    "                    # Gets the URL response so that the contents of the URL could be saved\n",
    "                    response = R.get(url)\n",
    "\n",
    "                    # Prints URLs just to see how it came out\n",
    "                    print(url)\n",
    "                    \n",
    "                    # For every id (idd) uniprot id we have, save the id's content into an XML file.\n",
    "                    # So let's say our id (idd) is 11\n",
    "                    # Then the xml file name would be 11.xml and the https://www.uniprot.org/uniprot/11.xml content ...\n",
    "                    # ... would be save into that 11.xml file\n",
    "                    #\n",
    "                    # Opens file\n",
    "                    output_file = open(f'{idd}.xml', 'w')\n",
    "                    # Writes to file\n",
    "                    output_file.write(response.text)\n",
    "                    # Closes the file\n",
    "                    output_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RNA name: mRNA_1\n",
      "Blast hit ID's: ['P52326', 'P26480', 'P52327', 'P0A2E3', 'P00579', 'O24744', 'P32001', 'P57163', 'Q89B10', 'Q83BB6', 'Q9PDM9', 'Q87DT7', 'Q8PG33', 'P43766', 'Q8P4H2', 'P52325', 'P33452', 'Q59753', 'Q2K619', 'P52324', 'D5AQI9', 'P17531', 'P0CZ15', 'Q1RKH7', 'Q92FZ8', 'Q4UJT1', 'P33451', 'Q68VQ5']\n",
      "P52326\n",
      "https://www.uniprot.org/uniprot/P52326.xml\n",
      "P26480\n",
      "https://www.uniprot.org/uniprot/P26480.xml\n",
      "P52327\n",
      "https://www.uniprot.org/uniprot/P52327.xml\n",
      "P0A2E3\n",
      "https://www.uniprot.org/uniprot/P0A2E3.xml\n",
      "P00579\n",
      "https://www.uniprot.org/uniprot/P00579.xml\n",
      "O24744\n",
      "https://www.uniprot.org/uniprot/O24744.xml\n",
      "P32001\n",
      "https://www.uniprot.org/uniprot/P32001.xml\n",
      "P57163\n",
      "https://www.uniprot.org/uniprot/P57163.xml\n",
      "Q89B10\n",
      "https://www.uniprot.org/uniprot/Q89B10.xml\n",
      "Q83BB6\n",
      "https://www.uniprot.org/uniprot/Q83BB6.xml\n",
      "Q9PDM9\n",
      "https://www.uniprot.org/uniprot/Q9PDM9.xml\n",
      "Q87DT7\n",
      "https://www.uniprot.org/uniprot/Q87DT7.xml\n",
      "Q8PG33\n",
      "https://www.uniprot.org/uniprot/Q8PG33.xml\n",
      "P43766\n",
      "https://www.uniprot.org/uniprot/P43766.xml\n",
      "Q8P4H2\n",
      "https://www.uniprot.org/uniprot/Q8P4H2.xml\n",
      "P52325\n",
      "https://www.uniprot.org/uniprot/P52325.xml\n",
      "P33452\n",
      "https://www.uniprot.org/uniprot/P33452.xml\n",
      "Q59753\n",
      "https://www.uniprot.org/uniprot/Q59753.xml\n",
      "Q2K619\n",
      "https://www.uniprot.org/uniprot/Q2K619.xml\n",
      "P52324\n",
      "https://www.uniprot.org/uniprot/P52324.xml\n",
      "D5AQI9\n",
      "https://www.uniprot.org/uniprot/D5AQI9.xml\n",
      "P17531\n",
      "https://www.uniprot.org/uniprot/P17531.xml\n",
      "P0CZ15\n",
      "https://www.uniprot.org/uniprot/P0CZ15.xml\n",
      "Q1RKH7\n",
      "https://www.uniprot.org/uniprot/Q1RKH7.xml\n",
      "Q92FZ8\n",
      "https://www.uniprot.org/uniprot/Q92FZ8.xml\n",
      "Q4UJT1\n",
      "https://www.uniprot.org/uniprot/Q4UJT1.xml\n",
      "P33451\n",
      "https://www.uniprot.org/uniprot/P33451.xml\n",
      "Q68VQ5\n",
      "https://www.uniprot.org/uniprot/Q68VQ5.xml\n",
      "RNA name: mRNA_2\n",
      "Blast hit ID's: []\n",
      "RNA name: mRNA_3\n",
      "Blast hit ID's: ['Q9RR29', 'D4GU70', 'P08075', 'A6VG23', 'A6UP85', 'A4FX98', 'Q58501', 'Q6LYB5', 'A6UUQ4', 'Q975F9', 'Q9Y725', 'Q6FRY2', 'P0C5I2', 'O93827', 'Q8BTZ7', 'Q9Y5P6', 'Q752H4', 'P41940', 'Q6BN12', 'O74624', 'Q2YDJ9', 'A2VD83', 'Q4U3E8']\n",
      "Q9RR29\n",
      "https://www.uniprot.org/uniprot/Q9RR29.xml\n",
      "D4GU70\n",
      "https://www.uniprot.org/uniprot/D4GU70.xml\n",
      "P08075\n",
      "https://www.uniprot.org/uniprot/P08075.xml\n",
      "A6VG23\n",
      "https://www.uniprot.org/uniprot/A6VG23.xml\n",
      "A6UP85\n",
      "https://www.uniprot.org/uniprot/A6UP85.xml\n",
      "A4FX98\n",
      "https://www.uniprot.org/uniprot/A4FX98.xml\n",
      "Q58501\n",
      "https://www.uniprot.org/uniprot/Q58501.xml\n",
      "Q6LYB5\n",
      "https://www.uniprot.org/uniprot/Q6LYB5.xml\n",
      "A6UUQ4\n",
      "https://www.uniprot.org/uniprot/A6UUQ4.xml\n",
      "Q975F9\n",
      "https://www.uniprot.org/uniprot/Q975F9.xml\n",
      "Q9Y725\n",
      "https://www.uniprot.org/uniprot/Q9Y725.xml\n",
      "Q6FRY2\n",
      "https://www.uniprot.org/uniprot/Q6FRY2.xml\n",
      "P0C5I2\n",
      "https://www.uniprot.org/uniprot/P0C5I2.xml\n",
      "O93827\n",
      "https://www.uniprot.org/uniprot/O93827.xml\n",
      "Q8BTZ7\n",
      "https://www.uniprot.org/uniprot/Q8BTZ7.xml\n",
      "Q9Y5P6\n",
      "https://www.uniprot.org/uniprot/Q9Y5P6.xml\n",
      "Q752H4\n",
      "https://www.uniprot.org/uniprot/Q752H4.xml\n",
      "P41940\n",
      "https://www.uniprot.org/uniprot/P41940.xml\n",
      "Q6BN12\n",
      "https://www.uniprot.org/uniprot/Q6BN12.xml\n",
      "O74624\n",
      "https://www.uniprot.org/uniprot/O74624.xml\n",
      "Q2YDJ9\n",
      "https://www.uniprot.org/uniprot/Q2YDJ9.xml\n",
      "A2VD83\n",
      "https://www.uniprot.org/uniprot/A2VD83.xml\n",
      "Q4U3E8\n",
      "https://www.uniprot.org/uniprot/Q4U3E8.xml\n",
      "RNA name: mRNA_4\n",
      "Blast hit ID's: ['P14738', 'Q2FE03', 'Q6G6H3', 'A7X6I5', 'Q6GDU5', 'Q2YW62', 'A0A0H2XKG3']\n",
      "P14738\n",
      "https://www.uniprot.org/uniprot/P14738.xml\n",
      "Q2FE03\n",
      "https://www.uniprot.org/uniprot/Q2FE03.xml\n",
      "Q6G6H3\n",
      "https://www.uniprot.org/uniprot/Q6G6H3.xml\n",
      "A7X6I5\n",
      "https://www.uniprot.org/uniprot/A7X6I5.xml\n",
      "Q6GDU5\n",
      "https://www.uniprot.org/uniprot/Q6GDU5.xml\n",
      "Q2YW62\n",
      "https://www.uniprot.org/uniprot/Q2YW62.xml\n",
      "A0A0H2XKG3\n",
      "https://www.uniprot.org/uniprot/A0A0H2XKG3.xml\n",
      "RNA name: mRNA_5\n",
      "Blast hit ID's: ['P08320', 'P39347', 'P37317', 'P37326', 'P32053', 'P76542']\n",
      "P08320\n",
      "https://www.uniprot.org/uniprot/P08320.xml\n",
      "P39347\n",
      "https://www.uniprot.org/uniprot/P39347.xml\n",
      "P37317\n",
      "https://www.uniprot.org/uniprot/P37317.xml\n",
      "P37326\n",
      "https://www.uniprot.org/uniprot/P37326.xml\n",
      "P32053\n",
      "https://www.uniprot.org/uniprot/P32053.xml\n",
      "P76542\n",
      "https://www.uniprot.org/uniprot/P76542.xml\n",
      "RNA name: mRNA_6\n",
      "Blast hit ID's: ['P69451', 'Q8XDR6', 'P63521', 'Q8ZES9', 'P46450', 'P94547', 'O07610', 'Q0DV32', 'P41636', 'Q42982', 'Q54P79', 'Q54P77', 'P14912', 'P14913', 'O24146', 'Q54P78']\n",
      "P69451\n",
      "https://www.uniprot.org/uniprot/P69451.xml\n",
      "Q8XDR6\n",
      "https://www.uniprot.org/uniprot/Q8XDR6.xml\n",
      "P63521\n",
      "https://www.uniprot.org/uniprot/P63521.xml\n",
      "Q8ZES9\n",
      "https://www.uniprot.org/uniprot/Q8ZES9.xml\n",
      "P46450\n",
      "https://www.uniprot.org/uniprot/P46450.xml\n",
      "P94547\n",
      "https://www.uniprot.org/uniprot/P94547.xml\n",
      "O07610\n",
      "https://www.uniprot.org/uniprot/O07610.xml\n",
      "Q0DV32\n",
      "https://www.uniprot.org/uniprot/Q0DV32.xml\n",
      "P41636\n",
      "https://www.uniprot.org/uniprot/P41636.xml\n",
      "Q42982\n",
      "https://www.uniprot.org/uniprot/Q42982.xml\n",
      "Q54P79\n",
      "https://www.uniprot.org/uniprot/Q54P79.xml\n",
      "Q54P77\n",
      "https://www.uniprot.org/uniprot/Q54P77.xml\n",
      "P14912\n",
      "https://www.uniprot.org/uniprot/P14912.xml\n",
      "P14913\n",
      "https://www.uniprot.org/uniprot/P14913.xml\n",
      "O24146\n",
      "https://www.uniprot.org/uniprot/O24146.xml\n",
      "Q54P78\n",
      "https://www.uniprot.org/uniprot/Q54P78.xml\n"
     ]
    }
   ],
   "source": [
    "## READS BLAST XML FILE AND SAVES BLAST IDS INTO AN XML FILE\n",
    "\n",
    "# Import biopython and python modules\n",
    "import requests as R\n",
    "from Bio import SeqIO\n",
    "import Bio.SearchIO as BSIO\n",
    "from Bio import SeqFeature\n",
    "from Bio.SeqFeature import SeqFeature\n",
    "from Bio import SeqFeature as sf\n",
    "import Bio.SeqFeature as BSF \n",
    "\n",
    "# Function for e-value that is less or equel to 10e-6\n",
    "def evalue(hsp):\n",
    "    return hsp.evalue <= 0.000001\n",
    "     \n",
    "# Function for span that is more or equel to 75 percent of the target sequence\n",
    "def span(hsp):\n",
    "    return (hsp.query_span / seq_len) * 100 >= 75\n",
    "\n",
    "# Function for percent identity that is more or equal to 30\n",
    "def identity(hsp):\n",
    "    return (hsp.ident_num / hsp.aln_span) * 100 >= 30\n",
    "\n",
    "# Main code that reads the file \n",
    "with open('2-blast.xml') as file:\n",
    "    for result in BSIO.parse(file, 'blast-xml'):\n",
    "        # Filteres all of the results that have that evalue()\n",
    "        filtered_result = result.hsp_filter(evalue)\n",
    "        # Sequnce lenghts with that evalue\n",
    "        seq_len = filtered_result.seq_len\n",
    "        # Filteres out the hits that have that evalue() filtered before but this time looks at the span()\n",
    "        results_forvalueNspan = filtered_result.hsp_filter(span)\n",
    "        # Filteres out the hits that have that evalue() and span() like filtered before but this time looks at the identity()\n",
    "        results_forAll = results_forvalueNspan.hsp_filter(identity)\n",
    "        #for hit in results_forAll:\n",
    "            #print(hit)\n",
    "        \n",
    "        # every mRNA1 name and so on from BLAST\n",
    "        rnaName = result.id\n",
    "        print(\"RNA name:\",rnaName)\n",
    "        \n",
    "        # Prints out all of the results that were filtered\n",
    "        db_xrefs_seq_ids = []\n",
    "        \n",
    "        # Gets all of the hit IDs withou th sp|| around it. Thats why we use accession\n",
    "        # Appends into a list\n",
    "        for hit in results_forAll:\n",
    "            idd = hit.accession\n",
    "            db_xrefs_seq_ids.append(idd)\n",
    "        print(\"Blast hit ID's:\",db_xrefs_seq_ids)\n",
    "        \n",
    "        # By the individual ID get the UniProt URLs and download the entries as XML files\n",
    "        for idd in db_xrefs_seq_ids:\n",
    "            # Prints ids just to check how they look\n",
    "            print(idd)\n",
    "        \n",
    "            # API address\n",
    "            base = \"https://www.uniprot.org/uniprot/%s.%s\"\n",
    "            \n",
    "            # Format\n",
    "            fmt = 'xml'\n",
    "            \n",
    "            # Adds the base url to idd and xml format by replaces the %s.%s\n",
    "            # It would be something like this if the id (idd) is 11:\n",
    "            # https://www.uniprot.org/uniprot/11.xml\n",
    "            url = base%(idd, fmt)\n",
    "            \n",
    "            # Gets the URL response so that the contents of the URL could be saved\n",
    "            response = R.get(url)\n",
    "            \n",
    "            # Prints URLs just to see how it came out\n",
    "            print(url)\n",
    "        \n",
    "            # For every id (idd) uniprot id we have, save the id's content into an XML file.\n",
    "            # So let's say our id (idd) is 11\n",
    "            # Then the xml file name would be 11.xml and the https://www.uniprot.org/uniprot/11.xml content ...\n",
    "            # ... would be save into that 11.xml file\n",
    "            #\n",
    "            # Opens file\n",
    "            output_file = open(f'{idd}.xml', 'w')\n",
    "            # Writes to file\n",
    "            output_file.write(response.text)\n",
    "            # Closes the file\n",
    "            output_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GO:0000001\n",
      "mitochondrion inheritance\n"
     ]
    }
   ],
   "source": [
    "import re as R\n",
    "import collections as C\n",
    "\n",
    "# matches stanza header\n",
    "r_header = R.compile(r'^\\[(\\w+)\\]$')\n",
    "# matches tag:value\n",
    "r_tag = R.compile(r'^(.*?): (.*)( ! .*)?$')\n",
    "# named tuple\n",
    "Term = C.namedtuple('Term', ['id', 'name'])\n",
    "\n",
    "def parse_go_obo(go_obo_file):\n",
    "    # result dictionary\n",
    "    go = {}\n",
    "    # current stanza data\n",
    "    head = ''\n",
    "    data = {}\n",
    "    # process one line at the time\n",
    "    with open(go_obo_file, 'rt') as f:\n",
    "        for line in f:\n",
    "            # strip newline\n",
    "            line = line.strip()\n",
    "            # skip over empty lines\n",
    "            if not line:\n",
    "                continue\n",
    "            # check if new stanza is encountered\n",
    "            header = r_header.match(line)\n",
    "            if header:\n",
    "                # flush current term if any\n",
    "                if head == 'Term' and len(data) == 2:\n",
    "                    term = Term(**data)\n",
    "                    go[term.id] = term\n",
    "                # start new stanza\n",
    "                head = header.group(1)\n",
    "                data = {}\n",
    "            else:\n",
    "                # collect data\n",
    "                tag, value, _ = r_tag.match(line).groups()\n",
    "                # store only needed tags\n",
    "                if tag in ['id', 'name']:\n",
    "                    data[tag] = value\n",
    "        # flush the last term\n",
    "        if head == 'Term' and len(data) == 2:\n",
    "            term = Term(**data)\n",
    "            go[term.id] = term\n",
    "    # return result\n",
    "    return go\n",
    "\n",
    "go = parse_go_obo('go.obo')\n",
    "\n",
    "# (just a brief check what is in dictionary)\n",
    "k = go['GO:0000001']\n",
    "for i in k:\n",
    "    print(i)\n",
    "#print(go['GO:0000001'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Protein:mRNA_1\n",
      "--------------\n",
      "GO:0001123\n",
      "transcription initiation from bacterial-type RNA polymerase promoter\n",
      "\n",
      "GO:0003677\n",
      "DNA binding\n",
      "\n",
      "GO:0003700\n",
      "DNA-binding transcription factor activity\n",
      "\n",
      "GO:0016987\n",
      "sigma factor activity\n",
      "\n",
      "\n",
      "Protein:mRNA_2\n",
      "--------------\n",
      "\n",
      "Protein:mRNA_3\n",
      "--------------\n",
      "\n",
      "Protein:mRNA_4\n",
      "--------------\n",
      "GO:0005576\n",
      "extracellular region\n",
      "\n",
      "GO:0005618\n",
      "cell wall\n",
      "\n",
      "GO:0007155\n",
      "cell adhesion\n",
      "\n",
      "GO:0009405\n",
      "pathogenesis\n",
      "\n",
      "GO:0016020\n",
      "membrane\n",
      "\n",
      "\n",
      "Protein:mRNA_5\n",
      "--------------\n",
      "GO:0044826\n",
      "viral genome integration into host DNA\n",
      "\n",
      "GO:0046718\n",
      "viral entry into host cell\n",
      "\n",
      "\n",
      "Protein:mRNA_6\n",
      "--------------\n",
      "GO:0005524\n",
      "ATP binding\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## GO TERM RETRIEVAL BASED ON OCCURANCE IN SEVERAL BLAST IDS\n",
    "\n",
    "from Bio import SeqIO\n",
    "import collections as C\n",
    "import requests as R\n",
    "import itertools\n",
    "import collections\n",
    "from collections import Counter\n",
    "import Bio.SeqFeature as BSF\n",
    "\n",
    "# Our genbank file\n",
    "proteins2 = list(SeqIO.parse(\"2-proteins.gb\",\"gb\"))\n",
    "#print(proteins2)\n",
    "\n",
    "# Skimming trough all of the protein information in the 2-proteins.gb file\n",
    "# Goes 1 by 1\n",
    "for protein in proteins2:\n",
    "    \n",
    "    print()\n",
    "    \n",
    "    # Name of protein\n",
    "    names = \"Protein:\" + protein.name\n",
    "    print(names)\n",
    "    \n",
    "    # Underline of protein name\n",
    "    print(len(names)*'-')\n",
    "    #print(protein)\n",
    "    \n",
    "    # Finds all of the IDs that are in genbank file 2-proteins with feature - BLAST\n",
    "    for f in protein.features:\n",
    "        # Check if the feature in genbank file is BLAST\n",
    "        if f.type == \"BLAST\":\n",
    "            # Check if qualifier is db_xref, which has our BLAST IDs.\n",
    "            if \"db_xref\" in f.qualifiers:\n",
    "                # Get all of those IDs\n",
    "                idss = f.qualifiers[\"db_xref\"]\n",
    "                #print(idss)\n",
    "                \n",
    "                # Lenght of how many BLAST IDs we have for every protein mRNA_\n",
    "                lenIDs = len(idss)\n",
    "                #print(lenID)\n",
    "                \n",
    "                # Empty list to store ALL GO terms from ALL of the xml files\n",
    "                # It's gonna be a list of lists like so: [[],[],...]\n",
    "                goTerms = []\n",
    "                \n",
    "                # For every ID in all genbank file BLAST IDs ...\n",
    "                for idd in idss:\n",
    "                    \n",
    "                    # Read the xml files which have the genbank BLAST IDs that we took to download our uniprot xml files before\n",
    "                    # So for every BLAST ID in the in genbank file we read an .xml file associated with that ID\n",
    "                    fullProtein = SeqIO.read(f'{idd}.xml','uniprot-xml')\n",
    "                    #print(fullProtein)\n",
    "    \n",
    "                    # Gets all of the GO IDs from ALL of the XML files we read before\n",
    "                    for ref in fullProtein.dbxrefs:\n",
    "                        # Cuts the ID because if not, it would start as GO:GO----\n",
    "                        goTerm = ref[3:]\n",
    "                        # If the ref (dbReference in xml file) starts with GO, append it to the list of lists\n",
    "                        if ref.startswith('GO:'):\n",
    "                            #print(goTerm)\n",
    "                            goTerms.append(goTerm)\n",
    "                #print(goTerms)\n",
    "                \n",
    "                # Count how many of every single term there is\n",
    "                # For example GO: 1 occurs 24 times, GO: 2 occurs 10 times and so on\n",
    "                dictt = Counter(goTerms)\n",
    "                #print(dictt)\n",
    "                \n",
    "                # Empty list for all of the Terms that we need\n",
    "                # All of them that occurs in all of the BLAST IDs we have for every protein mRNA_\n",
    "                AllTerms = []\n",
    "                \n",
    "                # Here we loop trough out dictt with the GO terms counted\n",
    "                # We get the second item of the dictionary: 'GO: 1': 7, so we get the 7\n",
    "                # And we look if the second item, for eg.: that 7, is equal to all of the BLAST IDs for a protein, the lenght of BLAST IDs\n",
    "                # item1 is our GO and item2 is our number (how many times it occurs)\n",
    "                # We append it to AllTerms list only those GO terms that occur in all of the BLAST IDs\n",
    "                # If we have 7 blast ids and we get GO that occurs 7, we append it\n",
    "                # If we got a GO that occured 6 times we dont, because it occured only in 6 BLAST IDs and not all 7\n",
    "                for item1, item2 in dictt.items():\n",
    "                    if item2 == lenIDs:\n",
    "                        AllTerms.append(item1)\n",
    "                #print(AllTerms)\n",
    "                \n",
    "                # Modified duplicate finder by georg --> https://stackoverflow.com/questions/9835762/how-do-i-find-the-duplicates-in-a-list-and-create-another-list-with-them\n",
    "                # Terms Gets only the duplicates in every list\n",
    "                # Not what we need so its incorrect\n",
    "                #Terms = [item for item, count in collections.Counter(goTerms).items() if count > 1]\n",
    "                #print(Terms)\n",
    "                \n",
    "                # For every term in Terms variable we get the human readable names of it using parse_go_obo() function\n",
    "                # And then print it out nicely\n",
    "                for everyGo in AllTerms:\n",
    "                    # go[] is the parser\n",
    "                    content = go[everyGo]\n",
    "                    for stuff in content:\n",
    "                        print(stuff)\n",
    "                    print()\n",
    "                \n",
    "                # Creation of GO feature for our protein sequences'\n",
    "                # Len vieno protein o ne visu proteins yra tas 1..616 salia BLAST\n",
    "                feature = BSF.SeqFeature(BSF.FeatureLocation(0,len(protein)), type='GO')\n",
    "                \n",
    "                # We add a qualifier 'db_xref' to the GO IDs (Terms) and add it to GO feature\n",
    "                feature.qualifiers['db_xref'] = AllTerms\n",
    "                \n",
    "                # In the same GO feature we add location/qualifier that is 'evidence'\n",
    "                feature.qualifiers['evidence'] = 'inferred by similarity'\n",
    "        \n",
    "                # Then for every protein (not the whole stack of proteins but just every individual protein/mRNA)in the genbank file \n",
    "                # we add those features that have a name GO on the left collum\n",
    "                # and we add all of the feature items, I mean all of those ids with the qualifier\n",
    "                protein.features.append(feature)\n",
    "                #print(features)\n",
    "                \n",
    "# Write to genbank file\n",
    "SeqIO.write(proteins2, \"3-proteins.gb\",\"gb\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b><u>Report/Disscussion</u></b><p>\n",
    "The whole idea was to get GO terms for our protein sequences and get the human readable names.<p>\n",
    "So firstly we need to download the UniProt files for every ID that I have gotten after running BLAST. I made two ways you can do that: from our already made genbank file and from our BLAST xml. The code downloads all of the UniProt files in .xml format for every BLAST ID. Then the code reads all of the .xml files coresponding to the protein (mRNA) it was assigned to in BLAST and gets the GO terms. Since the `for loop` loops and gets all of the .xml file names or the BLAST IDs to its corresponding mRNA, that means that all of the GO terms found are stored into a list. For exmaple: for mRNA_1 we have a list with GO terms from certain .xmls, then for mRNA_2 and so on. So my idea was to put ALL of the GO terms found in EVERY .xml file and just keep the the GO terms that occur in every .xml file. So every list [] of GO terms is put into a list of lists [[],[],...]. The it that list get `Counter()` function called upon and that function counts how many of times it occured is a value. Afte that, we check if the counted number (occurancen number or value) is equal to our number of BLAST IDs for a singal query/original protein. So if we for mRNA_1 we have 6 BLAST IDs then we get how many we have, and compare to the value of our new dictionary. If so, then that means, that those GO terms appear in every BLAST .xml file for that protein. Then we put it into a list. <br>\n",
    "To get the human readable names I need a `go.obo` file which is provided by the UniProt already. I used a parses that was made in the exercize 4, which parses the `go.obo`. So I just had to use the parser and parse it by the GO terms I have gotten.<p>\n",
    "We can seee that mRNA_2 does not have any GO terms since there were no BLAST IDs based on our criteria, and also mRNA_3 has no GO terms. mRNA_4 has the most GO terms, while mRNA_6 has the least. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4. Sequence alignment\n",
    "\n",
    "Finds the positions at which the sequences differ from consensus.\n",
    "\n",
    "a) For each target protein sequence in the task 1, aligns the target sequence with the sequences of the UniProt entries in the task 3. Saves the alignments to the file `3-alignment-X.fasta` in the FASTA format where `X` is a running number.\n",
    "\n",
    "b) Finds the positions in the alignments at which the target sequence is not identical to the consensus. Uses 60% consensus (i.e. the consensus exists if at least 60% of the aligned sequences contain the same residue). Ignores the non-consensus positions and the gaps in either sequence.\n",
    "\n",
    "c) Summarises discoveries by reporting the positions where the target protein sequence is not identical to the consensus sequence for each alignment.\n",
    "\n",
    "d) Updates the records of the target protein sequences by adding the non-identical positions. Uses a sequence feature with the type `mutation` and the qualifier keys `observed` and `consensus` to store the residues in the target sequence and the consensus sequence, respectively. Saves the updated records to the file `4-proteins.gb`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['P52326', 'P26480', 'P52327', 'P0A2E3', 'P00579', 'O24744', 'P32001', 'P57163', 'Q89B10', 'Q83BB6', 'Q9PDM9', 'Q87DT7', 'Q8PG33', 'P43766', 'Q8P4H2', 'P52325', 'P33452', 'Q59753', 'Q2K619', 'P52324', 'D5AQI9', 'P17531', 'P0CZ15', 'Q1RKH7', 'Q92FZ8', 'Q4UJT1', 'P33451', 'Q68VQ5']\n",
      "P52326\n",
      "https://www.uniprot.org/uniprot/P52326.fasta\n",
      "P26480\n",
      "https://www.uniprot.org/uniprot/P26480.fasta\n",
      "P52327\n",
      "https://www.uniprot.org/uniprot/P52327.fasta\n",
      "P0A2E3\n",
      "https://www.uniprot.org/uniprot/P0A2E3.fasta\n",
      "P00579\n",
      "https://www.uniprot.org/uniprot/P00579.fasta\n",
      "O24744\n",
      "https://www.uniprot.org/uniprot/O24744.fasta\n",
      "P32001\n",
      "https://www.uniprot.org/uniprot/P32001.fasta\n",
      "P57163\n",
      "https://www.uniprot.org/uniprot/P57163.fasta\n",
      "Q89B10\n",
      "https://www.uniprot.org/uniprot/Q89B10.fasta\n",
      "Q83BB6\n",
      "https://www.uniprot.org/uniprot/Q83BB6.fasta\n",
      "Q9PDM9\n",
      "https://www.uniprot.org/uniprot/Q9PDM9.fasta\n",
      "Q87DT7\n",
      "https://www.uniprot.org/uniprot/Q87DT7.fasta\n",
      "Q8PG33\n",
      "https://www.uniprot.org/uniprot/Q8PG33.fasta\n",
      "P43766\n",
      "https://www.uniprot.org/uniprot/P43766.fasta\n",
      "Q8P4H2\n",
      "https://www.uniprot.org/uniprot/Q8P4H2.fasta\n",
      "P52325\n",
      "https://www.uniprot.org/uniprot/P52325.fasta\n",
      "P33452\n",
      "https://www.uniprot.org/uniprot/P33452.fasta\n",
      "Q59753\n",
      "https://www.uniprot.org/uniprot/Q59753.fasta\n",
      "Q2K619\n",
      "https://www.uniprot.org/uniprot/Q2K619.fasta\n",
      "P52324\n",
      "https://www.uniprot.org/uniprot/P52324.fasta\n",
      "D5AQI9\n",
      "https://www.uniprot.org/uniprot/D5AQI9.fasta\n",
      "P17531\n",
      "https://www.uniprot.org/uniprot/P17531.fasta\n",
      "P0CZ15\n",
      "https://www.uniprot.org/uniprot/P0CZ15.fasta\n",
      "Q1RKH7\n",
      "https://www.uniprot.org/uniprot/Q1RKH7.fasta\n",
      "Q92FZ8\n",
      "https://www.uniprot.org/uniprot/Q92FZ8.fasta\n",
      "Q4UJT1\n",
      "https://www.uniprot.org/uniprot/Q4UJT1.fasta\n",
      "P33451\n",
      "https://www.uniprot.org/uniprot/P33451.fasta\n",
      "Q68VQ5\n",
      "https://www.uniprot.org/uniprot/Q68VQ5.fasta\n",
      "['Q9RR29', 'D4GU70', 'P08075', 'A6VG23', 'A6UP85', 'A4FX98', 'Q58501', 'Q6LYB5', 'A6UUQ4', 'Q975F9', 'Q9Y725', 'Q6FRY2', 'P0C5I2', 'O93827', 'Q8BTZ7', 'Q9Y5P6', 'Q752H4', 'P41940', 'Q6BN12', 'O74624', 'Q2YDJ9', 'A2VD83', 'Q4U3E8']\n",
      "Q9RR29\n",
      "https://www.uniprot.org/uniprot/Q9RR29.fasta\n",
      "D4GU70\n",
      "https://www.uniprot.org/uniprot/D4GU70.fasta\n",
      "P08075\n",
      "https://www.uniprot.org/uniprot/P08075.fasta\n",
      "A6VG23\n",
      "https://www.uniprot.org/uniprot/A6VG23.fasta\n",
      "A6UP85\n",
      "https://www.uniprot.org/uniprot/A6UP85.fasta\n",
      "A4FX98\n",
      "https://www.uniprot.org/uniprot/A4FX98.fasta\n",
      "Q58501\n",
      "https://www.uniprot.org/uniprot/Q58501.fasta\n",
      "Q6LYB5\n",
      "https://www.uniprot.org/uniprot/Q6LYB5.fasta\n",
      "A6UUQ4\n",
      "https://www.uniprot.org/uniprot/A6UUQ4.fasta\n",
      "Q975F9\n",
      "https://www.uniprot.org/uniprot/Q975F9.fasta\n",
      "Q9Y725\n",
      "https://www.uniprot.org/uniprot/Q9Y725.fasta\n",
      "Q6FRY2\n",
      "https://www.uniprot.org/uniprot/Q6FRY2.fasta\n",
      "P0C5I2\n",
      "https://www.uniprot.org/uniprot/P0C5I2.fasta\n",
      "O93827\n",
      "https://www.uniprot.org/uniprot/O93827.fasta\n",
      "Q8BTZ7\n",
      "https://www.uniprot.org/uniprot/Q8BTZ7.fasta\n",
      "Q9Y5P6\n",
      "https://www.uniprot.org/uniprot/Q9Y5P6.fasta\n",
      "Q752H4\n",
      "https://www.uniprot.org/uniprot/Q752H4.fasta\n",
      "P41940\n",
      "https://www.uniprot.org/uniprot/P41940.fasta\n",
      "Q6BN12\n",
      "https://www.uniprot.org/uniprot/Q6BN12.fasta\n",
      "O74624\n",
      "https://www.uniprot.org/uniprot/O74624.fasta\n",
      "Q2YDJ9\n",
      "https://www.uniprot.org/uniprot/Q2YDJ9.fasta\n",
      "A2VD83\n",
      "https://www.uniprot.org/uniprot/A2VD83.fasta\n",
      "Q4U3E8\n",
      "https://www.uniprot.org/uniprot/Q4U3E8.fasta\n",
      "['P14738', 'Q2FE03', 'Q6G6H3', 'A7X6I5', 'Q6GDU5', 'Q2YW62', 'A0A0H2XKG3']\n",
      "P14738\n",
      "https://www.uniprot.org/uniprot/P14738.fasta\n",
      "Q2FE03\n",
      "https://www.uniprot.org/uniprot/Q2FE03.fasta\n",
      "Q6G6H3\n",
      "https://www.uniprot.org/uniprot/Q6G6H3.fasta\n",
      "A7X6I5\n",
      "https://www.uniprot.org/uniprot/A7X6I5.fasta\n",
      "Q6GDU5\n",
      "https://www.uniprot.org/uniprot/Q6GDU5.fasta\n",
      "Q2YW62\n",
      "https://www.uniprot.org/uniprot/Q2YW62.fasta\n",
      "A0A0H2XKG3\n",
      "https://www.uniprot.org/uniprot/A0A0H2XKG3.fasta\n",
      "['P08320', 'P39347', 'P37317', 'P37326', 'P32053', 'P76542']\n",
      "P08320\n",
      "https://www.uniprot.org/uniprot/P08320.fasta\n",
      "P39347\n",
      "https://www.uniprot.org/uniprot/P39347.fasta\n",
      "P37317\n",
      "https://www.uniprot.org/uniprot/P37317.fasta\n",
      "P37326\n",
      "https://www.uniprot.org/uniprot/P37326.fasta\n",
      "P32053\n",
      "https://www.uniprot.org/uniprot/P32053.fasta\n",
      "P76542\n",
      "https://www.uniprot.org/uniprot/P76542.fasta\n",
      "['P69451', 'Q8XDR6', 'P63521', 'Q8ZES9', 'P46450', 'P94547', 'O07610', 'Q0DV32', 'P41636', 'Q42982', 'Q54P79', 'Q54P77', 'P14912', 'P14913', 'O24146', 'Q54P78']\n",
      "P69451\n",
      "https://www.uniprot.org/uniprot/P69451.fasta\n",
      "Q8XDR6\n",
      "https://www.uniprot.org/uniprot/Q8XDR6.fasta\n",
      "P63521\n",
      "https://www.uniprot.org/uniprot/P63521.fasta\n",
      "Q8ZES9\n",
      "https://www.uniprot.org/uniprot/Q8ZES9.fasta\n",
      "P46450\n",
      "https://www.uniprot.org/uniprot/P46450.fasta\n",
      "P94547\n",
      "https://www.uniprot.org/uniprot/P94547.fasta\n",
      "O07610\n",
      "https://www.uniprot.org/uniprot/O07610.fasta\n",
      "Q0DV32\n",
      "https://www.uniprot.org/uniprot/Q0DV32.fasta\n",
      "P41636\n",
      "https://www.uniprot.org/uniprot/P41636.fasta\n",
      "Q42982\n",
      "https://www.uniprot.org/uniprot/Q42982.fasta\n",
      "Q54P79\n",
      "https://www.uniprot.org/uniprot/Q54P79.fasta\n",
      "Q54P77\n",
      "https://www.uniprot.org/uniprot/Q54P77.fasta\n",
      "P14912\n",
      "https://www.uniprot.org/uniprot/P14912.fasta\n",
      "P14913\n",
      "https://www.uniprot.org/uniprot/P14913.fasta\n",
      "O24146\n",
      "https://www.uniprot.org/uniprot/O24146.fasta\n",
      "Q54P78\n",
      "https://www.uniprot.org/uniprot/Q54P78.fasta\n"
     ]
    }
   ],
   "source": [
    "## DOWNLOADS ALL OF THE SEQENCES OF BLAST IDS AND SAVES IT INTO THEIR SPECIFIC FILE\n",
    "\n",
    "# Import biopython and python modules\n",
    "from Bio import SeqIO\n",
    "import collections as C\n",
    "import requests as R\n",
    "import itertools\n",
    "import collections\n",
    "import Bio.SeqFeature as BSF\n",
    "from pathlib import Path\n",
    "\n",
    "# It prints out a list of IDS and then one by one the link of the ID.\n",
    "# It helps to indicate which one is being written and how much are left.\n",
    "\n",
    "proteins2 = list(SeqIO.parse(\"2-proteins.gb\",\"gb\"))\n",
    "#print(proteins2)\n",
    "\n",
    "for protein in proteins2:\n",
    "     \n",
    "    # Protein name like mRNA and so on\n",
    "    pname = protein.id\n",
    "    \n",
    "    # Original protein sequence\n",
    "    pseq = protein.seq\n",
    "    \n",
    "    # Opens file\n",
    "    output_file = open(f'{pname}.fasta', 'w')\n",
    "    \n",
    "    #print(protein)\n",
    "    # Finds all of the ID that are in genbank file 2-proteins-tryout with feature - BLAST\n",
    "    for f in protein.features:\n",
    "        if f.type == \"BLAST\":\n",
    "            if \"db_xref\" in f.qualifiers:\n",
    "                idss = f.qualifiers[\"db_xref\"]\n",
    "                print(idss)\n",
    "                \n",
    "                # By the individual ID get the UniProt URLs and download the entries as XML files\n",
    "                for idd in idss:\n",
    "                    # Prints ids just to check how they look\n",
    "                    print(idd)\n",
    "\n",
    "                    # API address\n",
    "                    base = \"https://www.uniprot.org/uniprot/%s.%s\"\n",
    "\n",
    "                    # Format\n",
    "                    fmt = 'fasta'\n",
    "\n",
    "                    # Adds the base url to idd and xml format by replaces the %s.%s\n",
    "                    # It would be something like this if the id (idd) is 11:\n",
    "                    # https://www.uniprot.org/uniprot/11.xml\n",
    "                    url = base%(idd, fmt)\n",
    "\n",
    "                    # Gets the URL response so that the contents of the URL could be saved\n",
    "                    response = R.get(url)\n",
    "\n",
    "                    # Prints URLs just to see how it came out\n",
    "                    print(url)\n",
    "\n",
    "                    # For every id (idd) uniprot id we have, save the id's content into an XML file.\n",
    "                    # So let's say our id (idd) is 11\n",
    "                    # Then the xml file name would be 11.xml and the https://www.uniprot.org/uniprot/11.xml content ...\n",
    "                    # ... would be save into that 11.xml file\n",
    "                    #\n",
    "                    # Writes to file\n",
    "                    output_file.write(response.text)\n",
    "                    \n",
    "    # Closes the file\n",
    "    output_file.close() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "## COMBINES THE ORIGINAL PROTEIN SEQUENCE TO ITS DEDICATED BLAST SEQUENCES\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "# For every our original mRNA, write it into a file with the same name.\n",
    "# Write the original protein sequence and then the BLAST hit sequences\n",
    "# For eg.: mRna_1 and then all BLAST ID corresponding to it, and then mRNA_2 and so on. \n",
    "for record in SeqIO.parse(\"1-proteins.fasta\", \"fasta\"):\n",
    "    \n",
    "    # Name of the protein\n",
    "    name = record.id\n",
    "    \n",
    "    # Read the .fasta files we downloaded in the previous cell\n",
    "    recordFasta = list(SeqIO.parse(f'{name}.fasta','fasta'))\n",
    "    #print(recordFasta)\n",
    "    \n",
    "    records = [record] + recordFasta\n",
    "    #print(records)\n",
    "    \n",
    "    # Write into a new file our original protein sequence and the BLAST hit sequences\n",
    "    SeqIO.write(records,f'{name}_new.fasta','fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class MuscleCommandline in module Bio.Align.Applications._Muscle:\n",
      "\n",
      "class MuscleCommandline(Bio.Application.AbstractCommandline)\n",
      " |  MuscleCommandline(cmd='muscle', **kwargs)\n",
      " |  \n",
      " |  Command line wrapper for the multiple alignment program MUSCLE.\n",
      " |  \n",
      " |  http://www.drive5.com/muscle/\n",
      " |  \n",
      " |  Notes\n",
      " |  -----\n",
      " |  Last checked against version: 3.7, briefly against 3.8\n",
      " |  \n",
      " |  References\n",
      " |  ----------\n",
      " |  Edgar, Robert C. (2004), MUSCLE: multiple sequence alignment with high\n",
      " |  accuracy and high throughput, Nucleic Acids Research 32(5), 1792-97.\n",
      " |  \n",
      " |  Edgar, R.C. (2004) MUSCLE: a multiple sequence alignment method with\n",
      " |  reduced time and space complexity. BMC Bioinformatics 5(1): 113.\n",
      " |  \n",
      " |  Examples\n",
      " |  --------\n",
      " |  >>> from Bio.Align.Applications import MuscleCommandline\n",
      " |  >>> muscle_exe = r\"C:\\Program Files\\Aligments\\muscle3.8.31_i86win32.exe\"\n",
      " |  >>> in_file = r\"C:\\My Documents\\unaligned.fasta\"\n",
      " |  >>> out_file = r\"C:\\My Documents\\aligned.fasta\"\n",
      " |  >>> muscle_cline = MuscleCommandline(muscle_exe, input=in_file, out=out_file)\n",
      " |  >>> print(muscle_cline)\n",
      " |  \"C:\\Program Files\\Aligments\\muscle3.8.31_i86win32.exe\" -in \"C:\\My Documents\\unaligned.fasta\" -out \"C:\\My Documents\\aligned.fasta\"\n",
      " |  \n",
      " |  You would typically run the command line with muscle_cline() or via\n",
      " |  the Python subprocess module, as described in the Biopython tutorial.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      MuscleCommandline\n",
      " |      Bio.Application.AbstractCommandline\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, cmd='muscle', **kwargs)\n",
      " |      Initialize the class.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  __call__(self, stdin=None, stdout=True, stderr=True, cwd=None, env=None)\n",
      " |      Execute command, wait for it to finish, return (stdout, stderr).\n",
      " |      \n",
      " |      Runs the command line tool and waits for it to finish. If it returns\n",
      " |      a non-zero error level, an exception is raised. Otherwise two strings\n",
      " |      are returned containing stdout and stderr.\n",
      " |      \n",
      " |      The optional stdin argument should be a string of data which will be\n",
      " |      passed to the tool as standard input.\n",
      " |      \n",
      " |      The optional stdout and stderr argument may be filenames (string),\n",
      " |      but otherwise are treated as a booleans, and control if the output\n",
      " |      should be captured as strings (True, default), or ignored by sending\n",
      " |      it to /dev/null to avoid wasting memory (False). If sent to a file\n",
      " |      or ignored, then empty string(s) are returned.\n",
      " |      \n",
      " |      The optional cwd argument is a string giving the working directory\n",
      " |      to run the command from. See Python's subprocess module documentation\n",
      " |      for more details.\n",
      " |      \n",
      " |      The optional env argument is a dictionary setting the environment\n",
      " |      variables to be used in the new process. By default the current\n",
      " |      process' environment variables are used. See Python's subprocess\n",
      " |      module documentation for more details.\n",
      " |      \n",
      " |      Default example usage::\n",
      " |      \n",
      " |          from Bio.Emboss.Applications import WaterCommandline\n",
      " |          water_cmd = WaterCommandline(gapopen=10, gapextend=0.5,\n",
      " |                                       stdout=True, auto=True,\n",
      " |                                       asequence=\"a.fasta\", bsequence=\"b.fasta\")\n",
      " |          print(\"About to run: %s\" % water_cmd)\n",
      " |          std_output, err_output = water_cmd()\n",
      " |      \n",
      " |      This functionality is similar to subprocess.check_output() added in\n",
      " |      Python 2.7. In general if you require more control over running the\n",
      " |      command, use subprocess directly.\n",
      " |      \n",
      " |      As of Biopython 1.56, when the program called returns a non-zero error\n",
      " |      level, a custom ApplicationError exception is raised. This includes\n",
      " |      any stdout and stderr strings captured as attributes of the exception\n",
      " |      object, since they may be useful for diagnosing what went wrong.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return a representation of the command line object for debugging.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5)\n",
      " |      >>> cline.asequence = \"asis:ACCCGGGCGCGGT\"\n",
      " |      >>> cline.bsequence = \"asis:ACCCGAGCGCGGT\"\n",
      " |      >>> cline.outfile = \"temp_water.txt\"\n",
      " |      >>> print(cline)\n",
      " |      water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5\n",
      " |      >>> cline\n",
      " |      WaterCommandline(cmd='water', outfile='temp_water.txt', asequence='asis:ACCCGGGCGCGGT', bsequence='asis:ACCCGAGCGCGGT', gapopen=10, gapextend=0.5)\n",
      " |  \n",
      " |  __setattr__(self, name, value)\n",
      " |      Set attribute name to value (PRIVATE).\n",
      " |      \n",
      " |      This code implements a workaround for a user interface issue.\n",
      " |      Without this __setattr__ attribute-based assignment of parameters\n",
      " |      will silently accept invalid parameters, leading to known instances\n",
      " |      of the user assuming that parameters for the application are set,\n",
      " |      when they are not.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5, stdout=True)\n",
      " |      >>> cline.asequence = \"a.fasta\"\n",
      " |      >>> cline.bsequence = \"b.fasta\"\n",
      " |      >>> cline.csequence = \"c.fasta\"\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      ValueError: Option name csequence was not found.\n",
      " |      >>> print(cline)\n",
      " |      water -stdout -asequence=a.fasta -bsequence=b.fasta -gapopen=10 -gapextend=0.5\n",
      " |      \n",
      " |      This workaround uses a whitelist of object attributes, and sets the\n",
      " |      object attribute list as normal, for these.  Other attributes are\n",
      " |      assumed to be parameters, and passed to the self.set_parameter method\n",
      " |      for validation and assignment.\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Make the commandline string with the currently set options.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5)\n",
      " |      >>> cline.asequence = \"asis:ACCCGGGCGCGGT\"\n",
      " |      >>> cline.bsequence = \"asis:ACCCGAGCGCGGT\"\n",
      " |      >>> cline.outfile = \"temp_water.txt\"\n",
      " |      >>> print(cline)\n",
      " |      water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5\n",
      " |      >>> str(cline)\n",
      " |      'water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5'\n",
      " |  \n",
      " |  set_parameter(self, name, value=None)\n",
      " |      Set a commandline option for a program (OBSOLETE).\n",
      " |      \n",
      " |      Every parameter is available via a property and as a named\n",
      " |      keyword when creating the instance. Using either of these is\n",
      " |      preferred to this legacy set_parameter method which is now\n",
      " |      OBSOLETE, and likely to be DEPRECATED and later REMOVED in\n",
      " |      future releases.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  parameters = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# MUSCLE help\n",
    "\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from Bio import AlignIO\n",
    "\n",
    "help(MuscleCommandline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "mRNA_1\n",
      "------\n",
      "SingleLetterAlphabet() alignment with 29 rows and 821 columns\n",
      "--------------------------------------------...--- sp|Q1RKH7|RPOD_RICBR\n",
      "--------------------------------------------...--- sp|P33451|RPOD_RICPR\n",
      "--------------------------------------------...--- sp|Q68VQ5|RPOD_RICTY\n",
      "--------------------------------------------...--- sp|Q92FZ8|RPOD_RICCN\n",
      "--------------------------------------------...--- sp|Q4UJT1|RPOD_RICFE\n",
      "--------------------------------------------...--- sp|D5AQI9|RPOD_RHOCB\n",
      "--------------------------------------------...--- sp|P0CZ15|RPOD_RHOCA\n",
      "--------------------------------------------...--- sp|P52324|RPOD_CAUVC\n",
      "--------------------------------------------...--- sp|P33452|RPOD_AGRFC\n",
      "--------------------------------------------...--- sp|Q59753|RPOD_RHIME\n",
      "--------------------------------------------...--- sp|Q2K619|RPOD_RHIEC\n",
      "MPTQKPSKASVKPTKKKVDPVIRKKKAPDGAAAKGAKTGAEEKE...--- sp|P17531|RPOD_MYXXA\n",
      "--------------------------------------------...SKL sp|P52325|RPOD_NEIGO\n",
      "---------------------MPSRKKKSSVTKTKKKTTGKKAR...--- sp|Q83BB6|RPOD_COXBU\n",
      "--------------------------------------------...--- sp|Q9PDM9|RPOD_XYLFA\n",
      "--------------------------------------------...--- sp|Q87DT7|RPOD_XYLFT\n",
      "--------------------------------------------...--- sp|Q8PG33|RPOD_XANAC\n",
      "--------------------------------------------...--- sp|Q8P4H2|RPOD_XANCP\n",
      "...\n",
      "--------------------------------------------...--- sp|P57163|RPOD_BUCAI\n",
      "\n",
      "mRNA_2\n",
      "------\n",
      "SingleLetterAlphabet() alignment with 1 rows and 372 columns\n",
      "MTVTPERSTYTPPGRLGDESSGPRTDPRFSPAMVEALATFGLDE...RRP mRNA_2\n",
      "\n",
      "mRNA_3\n",
      "------\n",
      "SingleLetterAlphabet() alignment with 24 rows and 442 columns\n",
      "MKALILVGGYGTRLRPLTLSVPKPLVDFCNKPILLHQVEALVKA...--- sp|A2VD83|GMPBB_XENLA\n",
      "MKALILVGGYGTRLRPLTLSIPKPLADFCNKPILLHQVEALAAA...--- sp|Q2YDJ9|GMPPB_BOVIN\n",
      "MKALILVGGYGTRLRPLTLSIPKPLVDFCNKPILLHQVEALASA...--- sp|P0C5I2|GMPPB_PIG\n",
      "MKALILVGGYGTRLRPLTLSTPKPLVDFCNKPILLHQVEALAAA...--- sp|Q8BTZ7|GMPPB_MOUSE\n",
      "MKALILVGGYGTRLRPLTLSTPKPLVDFCNKPILLHQVEALAAA...--- sp|Q9Y5P6|GMPPB_HUMAN\n",
      "MKGLILVGGFGTRLRPLTLTLPKPLVEFCNKPMIVHQIEALVAA...--- sp|O74624|MPG1_HYPJE\n",
      "MKALILVGGFGTRLRPLTLTLPKPLVEFGNRPMILHQVESLAAA...--- sp|Q4U3E8|MPG1_ASPFU\n",
      "MKGLILVGGYGTRLRPLTLTLPKPLVEFGNRPMILHQIEALAAA...--- sp|O93827|MPG1_CANAL\n",
      "MKGLILVGGYGTRLRPLTLTLPKPLVEFGNRPMILHQIEALAKA...--- sp|Q6BN12|MPG1_DEBHA\n",
      "MKGLILVGGYGTRLRPLTLSVPKPLVEFCNRPMILHQIEALAEA...--- sp|Q6FRY2|MPG12_CANGA\n",
      "MKGLILVGGYGTRLRPLTLTVPKPLVEFGNRPMILHQIEALANA...--- sp|P41940|MPG1_YEAST\n",
      "MKGLILVGGYGTRLRPLTLTVPKPLVEFGNRPMILHQIEALANA...--- sp|Q9Y725|MPG11_CANGA\n",
      "MKGLILVGGYGTRLRPLTLTVPKPLVEFCNRPMILHQIEALAAA...--- sp|Q752H4|MPG1_ASHGO\n",
      "MKAFILAAGSGERLEPITHTRPKAFVPILSKPLIEYQIEYLRKC...--- sp|Q975F9|S1PNA_SULTO\n",
      "MDAVILCAGSGTRLYPITENRPKPMIPIAGKPILEHIIEKIENH...KRN sp|A6UP85|GLMU_METVS\n",
      "MDAIILCAGKGTRLHPITESRPKPMIPIAGKPIIEHIIEKIENH...KRN sp|Q6LYB5|GLMU_METMP\n",
      "MDAIILCAGKGTRLYPITESRPKPMIPIAGKPILEHIIEKIENY...KRN sp|A6VG23|GLMU_METM7\n",
      "MDAIILCAGKGTRLHPITESRPKPMIPIAGKPILEHIIEKIENH...KRN sp|A4FX98|GLMU_METM5\n",
      "...\n",
      "MKALVLAGGTGTRLRPITHTSAKQLVPVANKPVLFYGLEAIRAA...--- sp|P08075|RMLA_STRGR\n",
      "\n",
      "mRNA_4\n",
      "------\n",
      "SingleLetterAlphabet() alignment with 8 rows and 1053 columns\n",
      "MKSNLRYGIRKHKLGAASVFLGTMIVVGMGQEKEAAASEQNNTT...HKA sp|A0A0H2XKG3|FNBB_STAA3\n",
      "MKNNLRYGIRKHKLGAASVFLGTMIVVGMGQDKEAAASEQKTTT...HKA sp|Q2FE03|FNBA_STAA3\n",
      "-----------------------MIVVGMGQDKEAAASEQKTTT...HKA mRNA_4\n",
      "MKNNLRYGIRKHKLGAASVFLGTMIVVGMGQDKEAAASEQKTTT...HKA sp|P14738|FNBA_STAA8\n",
      "MKNNLRYGIRKHKLGAASVFLGTMIVIGMGQDKEAAASEQKTTT...HKA sp|Q6GDU5|FNBA_STAAR\n",
      "MKNNLRYGIRKHKLGAASVFLGTMIVVGMGQDKEAATSEQKTTT...HKA sp|Q2YW62|FNBA_STAAB\n",
      "MKNNLRYGIRKHKLGAASVFLGTMIVVGMGQDKEAAASEQKTTT...NKA sp|Q6G6H3|FNBA_STAAS\n",
      "MKNNLRYGIRKHKLGAASVFLGTMIVVGMGQDKEAAASEQKTTT...NKA sp|A7X6I5|FNBA_STAA1\n",
      "\n",
      "mRNA_5\n",
      "------\n",
      "SingleLetterAlphabet() alignment with 7 rows and 448 columns\n",
      "MARKTKPLTDTEIKAAKPKDADYQLYDG--DGLTLLIKSSGSKL...--- sp|P32053|INTA_ECOLI\n",
      "-----MPLTDLKV---RSLTTPGRYADG--GGMYLQITASGGRY...--- mRNA_5\n",
      "-MCPQMKLNARQVETAKPKDKTYKMADG--GGLYLEVSAKGSKY...SVK sp|P08320|VINT_BPP4\n",
      "--------------------------------MHLLVHPNGSKY...--- sp|P39347|INTB_ECOLI\n",
      "--MSKTPLTAKAIDAAQPQDKPYKLTDSLTPGLFLLVHPNGSKY...--- sp|P76542|INTZ_ECOLI\n",
      "------MLTVKQIEAAKPKEKPYRLLDG--NGLYLYVPVSGKKV...--- sp|P37317|VINT_BPSFV\n",
      "------MLTVKQIEAAKPKEKPYRLLDG--NGLYLYVPVSGKKV...--- sp|P37326|INTS_ECOLI\n",
      "\n",
      "mRNA_6\n",
      "------\n",
      "SingleLetterAlphabet() alignment with 17 rows and 605 columns\n",
      "-----------------MIRIINGQTYFTSKYP---NIIIPE--...--- sp|Q54P79|4CL3_DICDI\n",
      "-----------------MIIKVKGQTYFTSKYP---NIIIPE--...--- sp|Q54P77|4CL1_DICDI\n",
      "-----------------MIRMIKGQIYFTSKYP---NIIIPE--...--- sp|Q54P78|4CL2_DICDI\n",
      "-----------MASASVPAAGYGADGVYRSLRP---PAPVASDP...--- sp|Q0DV32|4CLL1_ORYSJ\n",
      "MITVAAPEAQPQVAAAVDEAPPEAVTVFRSKLP---DIDIPS--...--- sp|Q42982|4CL2_ORYSJ\n",
      "----------------MANGIKKVEHLYRSKLP---DIEISD--...--- sp|P41636|4CL_PINTA\n",
      "----------------MEKDTKQVDIIFRSKLP---DIYIPN--...--- sp|O24146|4CL2_TOBAC\n",
      "---------------MGDCVAPKEDLIFRSKLP---DIYIPK--...K-- sp|P14912|4CL1_PETCR\n",
      "---------------MGDCVAPKEDLIFRSKLP---DIYIPK--...K-- sp|P14913|4CL2_PETCR\n",
      "-----------------------MEKIWFQNYPKGSEKFLDTSK...H-- sp|P46450|LCFA_HAEIN\n",
      "-----------------------MEKVWLKRYPADVPAEIDPDR...DAA sp|Q8ZES9|LCFA_YERPE\n",
      "-MLTACISFGVAMTTNTHFRGEELKKVWLNRYPADVPTEINPDR...--- mRNA_6\n",
      "-----------------------MKKVWLNRYPADVPAEINPDR...KA- sp|P63521|LCFA_SALTY\n",
      "-----------------------MKKVWLNRYPADVPTEINPDR...KA- sp|P69451|LCFA_ECOLI\n",
      "-----------------------MKKVWLNRYPADVPTEINPDR...KA- sp|Q8XDR6|LCFA_ECO57\n",
      "---------------------MQSQKPWLAEYPNDIPHELPL-P...K-- sp|P94547|LCFA_BACSU\n",
      "--------------------------------------------...--- sp|O07610|LCFB_BACSU\n"
     ]
    }
   ],
   "source": [
    "## MSA AND ITS FILE CREATION\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "import Bio.Align.Applications\n",
    "from Bio import AlignIO\n",
    "\n",
    "# Putting all of the genbank file into the list\n",
    "proteins3 = list(SeqIO.parse(\"3-proteins.gb\",\"gb\"))\n",
    "\n",
    "# So for every protein in the 3-proteins.gb\n",
    "for protein in proteins3:\n",
    "    print()\n",
    "    # Gets the name of the protein\n",
    "    # So smth like mRNA_1 and so on ...\n",
    "    name = protein.name\n",
    "    print(name)\n",
    "    \n",
    "    print(len(name)*'-')\n",
    "          \n",
    "    # MUSCLE.exe domain that needs to be specified\n",
    "    muscleExe = r\"muscle3.8.31_i86win32.exe\"\n",
    "\n",
    "    try:\n",
    "        from StringIO import StringIO\n",
    "    except ImportError:\n",
    "        from io import StringIO\n",
    "\n",
    "    # Muscle command line where it asks a FASTA line all of the sequence to be aligned\n",
    "    # Output is the alignment\n",
    "    muscle_cline = MuscleCommandline(muscleExe, input = f'{name}_new.fasta', out = f'3-alignment-{name}.fasta')\n",
    "    # Run command line witht writing the output file\n",
    "    # Default format of MUSCLE is fasta\n",
    "    \n",
    "    # This line is nesecarry, if no such line, the program would crash with an error\n",
    "    # The result is in stdout\n",
    "    stdout, stderr = muscle_cline()\n",
    "    \n",
    "    # Prints the results\n",
    "    print(AlignIO.read(f'3-alignment-{name}.fasta', \"fasta\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. COMAPARED WITH THE TARGET PROTEIN SEQUENCE THAT IS ALIGNED IN THE ALIGNMENT AND SAVE RESIDUES IN FILE<p>\n",
    "2. COMPARED WITH THE ORIGINAL UNALIGNED SEQUENCE AND SAVE RESIDUES IN FILE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Different from consensus sequence:\n",
      "\n",
      " mRNA_1\n",
      "18 R K\n",
      "42 V I\n",
      "70 D E\n",
      "163 L I\n",
      "198 D E\n",
      "328 Y W\n",
      "486 G A\n",
      "529 T N\n",
      "551 S A\n",
      "555 A P\n",
      "604 T S\n",
      "\n",
      " mRNA_2\n",
      "\n",
      " mRNA_3\n",
      "2 G A\n",
      "7 A G\n",
      "24 A L\n",
      "25 I V\n",
      "29 G N\n",
      "30 R K\n",
      "35 Y H\n",
      "50 V L\n",
      "54 P Y\n",
      "58 R E\n",
      "65 E K\n",
      "69 V I\n",
      "78 Q L\n",
      "80 L T\n",
      "94 S D\n",
      "98 L V\n",
      "99 Y L\n",
      "100 L N\n",
      "103 N V\n",
      "104 L I\n",
      "127 V T\n",
      "128 R K\n",
      "133 R S\n",
      "145 V I\n",
      "158 D N\n",
      "160 A I\n",
      "161 V N\n",
      "167 F L\n",
      "175 V I\n",
      "176 R E\n",
      "202 V L\n",
      "207 V L\n",
      "217 H P\n",
      "265 A V\n",
      "273 V I\n",
      "278 V I\n",
      "292 R G\n",
      "295 V I\n",
      "301 E L\n",
      "306 V I\n",
      "310 R S\n",
      "325 K G\n",
      "331 P R\n",
      "337 I V\n",
      "342 S V\n",
      "\n",
      " mRNA_4\n",
      "317 I V\n",
      "337 R K\n",
      "343 F Y\n",
      "361 M T\n",
      "374 R K\n",
      "376 F Y\n",
      "385 I L\n",
      "401 E D\n",
      "431 V I\n",
      "453 H Y\n",
      "485 E G\n",
      "526 S T\n",
      "527 T P\n",
      "539 G E\n",
      "679 K E\n",
      "699 D G\n",
      "815 S N\n",
      "\n",
      " mRNA_5\n",
      "9 S P\n",
      "14 G Y\n",
      "22 M L\n",
      "26 I V\n",
      "32 R K\n",
      "56 G T\n",
      "57 I L\n",
      "75 E I\n",
      "108 A K\n",
      "112 E S\n",
      "116 L A\n",
      "144 V L\n",
      "145 M L\n",
      "174 G A\n",
      "184 T P\n",
      "193 I A\n",
      "203 A P\n",
      "243 V L\n",
      "253 L F\n",
      "262 P A\n",
      "278 E Q\n",
      "283 I L\n",
      "299 S G\n",
      "312 F V\n",
      "327 A G\n",
      "374 V R\n",
      "378 N Q\n",
      "382 Q D\n",
      "385 A D\n",
      "\n",
      " mRNA_6\n",
      "59 R D\n",
      "71 D T\n",
      "224 L I\n",
      "281 Y F\n",
      "307 R F\n",
      "328 N P\n",
      "356 M A\n",
      "466 E D\n",
      "490 R E\n",
      "492 K E\n",
      "495 S L\n",
      "498 M H\n",
      "499 V P\n",
      "504 S A\n",
      "506 L V\n",
      "507 L V\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 1. COMAPARED WITH THE TARGET PROTEIN SEQUENCE THAT IS ALIGNED IN THE ALIGNMENT AND SAVE RESIDUES IN FILE\n",
    "\n",
    "from Bio import AlignIO\n",
    "from Bio import SeqIO\n",
    "from Bio.Align import AlignInfo # To generate a trivial 'dumb' consensus\n",
    "from Bio.Align.AlignInfo import SummaryInfo\n",
    "## load AlignInfo modules and SummaryInfo submodules\n",
    "import Bio.SeqFeature as BSF\n",
    "import re\n",
    "\n",
    "try:\n",
    "    from itertools import izip as zip\n",
    "except ImportError: # will be 3.x series\n",
    "    pass\n",
    "\n",
    "# Putting all of the genbank file into the list\n",
    "proteins3 = list(SeqIO.parse(\"3-proteins.gb\",\"gb\"))\n",
    "\n",
    "print(\"Different from consensus sequence:\")\n",
    "\n",
    "for protein in proteins3:\n",
    "    \n",
    "    # Name of the protein\n",
    "    name = protein.id\n",
    "    print(\"\\n\",name)\n",
    "    \n",
    "    # Protein sequence\n",
    "    seq = protein.seq\n",
    "    #print(seq)\n",
    "    \n",
    "    # Alignment files being read base on protein name\n",
    "    align = AlignIO.read(f'3-alignment-{name}.fasta', \"fasta\")\n",
    "    #print(align)\n",
    "\n",
    "    # Summary of the alignment\n",
    "    summary = SummaryInfo(align)\n",
    "\n",
    "    # Create gap consensus sequence\n",
    "    dumb_cons = summary.gap_consensus(require_multiple=1, threshold=0.6)\n",
    "    \n",
    "    # For every sequence in the alignment\n",
    "    for seq in align:\n",
    "        \n",
    "        index = 0\n",
    "        \n",
    "        # Is there a sequence that starts with mRNA_\n",
    "        if re.search(\"mRNA_\", seq.id):\n",
    "            ogProt = seq.seq\n",
    "            \n",
    "            # Comparison of gap consensus sequence with out target protein\n",
    "            # Results are our positions that do not match\n",
    "            for ch1, ch2 in zip(ogProt, dumb_cons):\n",
    "\n",
    "                if ch1 != ch2 and ch1 not in \"-X\" and ch2 not in \"-X\":\n",
    "\n",
    "                    # Creation of GO feature for our protein sequences'\n",
    "                    # Len vieno protein o ne visu proteins yra tas 1..616 salia BLAST\n",
    "                    feature = BSF.SeqFeature(BSF.FeatureLocation(index, index + 1), type='mutation')\n",
    "\n",
    "                    # We add a qualifier 'db_xref' to the GO IDs (Terms) and add it to BLAST feature\n",
    "                    feature.qualifiers['observed'] = ch1\n",
    "\n",
    "                    # In the same GO feature we add location/qualifier that is 'evidence'\n",
    "                    feature.qualifiers['consensus'] = ch2\n",
    "\n",
    "                    # Then for every protein (not the whole stack of proteins but just every individual protein/mRNA)in the genbank file \n",
    "                    # we add those features that have a name GO on the left collum\n",
    "                    # and we add all of the feature items, I mean all of those ids with the qualifier\n",
    "                    protein.features.append(feature)\n",
    "                  \n",
    "                    print(index, ch1, ch2)\n",
    "                    \n",
    "                # Ignoring the gaps; not counting in while counting all of the residues\n",
    "                if ch1 != \"-\":\n",
    "                    index += 1\n",
    "                            \n",
    "# Write to genbank file\n",
    "SeqIO.write(proteins3, \"4-proteins.gb\",\"gb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "mRNA_1\n",
      "[18:19]\n",
      "[42:43]\n",
      "[70:71]\n",
      "[163:164]\n",
      "[198:199]\n",
      "[328:329]\n",
      "[486:487]\n",
      "[529:530]\n",
      "[551:552]\n",
      "[555:556]\n",
      "[604:605]\n",
      "\n",
      "mRNA_2\n",
      "\n",
      "mRNA_3\n",
      "[2:3]\n",
      "[7:8]\n",
      "[24:25]\n",
      "[25:26]\n",
      "[29:30]\n",
      "[30:31]\n",
      "[35:36]\n",
      "[50:51]\n",
      "[54:55]\n",
      "[58:59]\n",
      "[65:66]\n",
      "[69:70]\n",
      "[78:79]\n",
      "[80:81]\n",
      "[94:95]\n",
      "[98:99]\n",
      "[99:100]\n",
      "[100:101]\n",
      "[103:104]\n",
      "[104:105]\n",
      "[127:128]\n",
      "[128:129]\n",
      "[133:134]\n",
      "[145:146]\n",
      "[158:159]\n",
      "[160:161]\n",
      "[161:162]\n",
      "[167:168]\n",
      "[175:176]\n",
      "[176:177]\n",
      "[202:203]\n",
      "[207:208]\n",
      "[217:218]\n",
      "[265:266]\n",
      "[273:274]\n",
      "[278:279]\n",
      "[292:293]\n",
      "[295:296]\n",
      "[301:302]\n",
      "[306:307]\n",
      "[310:311]\n",
      "[325:326]\n",
      "[331:332]\n",
      "[337:338]\n",
      "[342:343]\n",
      "\n",
      "mRNA_4\n",
      "[317:318]\n",
      "[337:338]\n",
      "[343:344]\n",
      "[361:362]\n",
      "[374:375]\n",
      "[376:377]\n",
      "[385:386]\n",
      "[401:402]\n",
      "[431:432]\n",
      "[453:454]\n",
      "[485:486]\n",
      "[526:527]\n",
      "[527:528]\n",
      "[539:540]\n",
      "[679:680]\n",
      "[699:700]\n",
      "[815:816]\n",
      "\n",
      "mRNA_5\n",
      "[9:10]\n",
      "[14:15]\n",
      "[22:23]\n",
      "[26:27]\n",
      "[32:33]\n",
      "[56:57]\n",
      "[57:58]\n",
      "[75:76]\n",
      "[108:109]\n",
      "[112:113]\n",
      "[116:117]\n",
      "[144:145]\n",
      "[145:146]\n",
      "[174:175]\n",
      "[184:185]\n",
      "[193:194]\n",
      "[203:204]\n",
      "[243:244]\n",
      "[253:254]\n",
      "[262:263]\n",
      "[278:279]\n",
      "[283:284]\n",
      "[299:300]\n",
      "[312:313]\n",
      "[327:328]\n",
      "[374:375]\n",
      "[378:379]\n",
      "[382:383]\n",
      "[385:386]\n",
      "\n",
      "mRNA_6\n",
      "[59:60]\n",
      "[71:72]\n",
      "[224:225]\n",
      "[281:282]\n",
      "[307:308]\n",
      "[328:329]\n",
      "[356:357]\n",
      "[466:467]\n",
      "[490:491]\n",
      "[492:493]\n",
      "[495:496]\n",
      "[498:499]\n",
      "[499:500]\n",
      "[504:505]\n",
      "[506:507]\n",
      "[507:508]\n"
     ]
    }
   ],
   "source": [
    "# 1. CONTINUATION\n",
    "\n",
    "from Bio import SeqIO\n",
    "import Bio.Entrez as BE\n",
    "import Bio.SearchIO as BSIO\n",
    "from Bio import SeqFeature\n",
    "from Bio.SeqFeature import SeqFeature\n",
    "from Bio import SeqFeature as sf\n",
    "import Bio.SeqFeature as BSF\n",
    "\n",
    "# Our updated file\n",
    "seq = list(SeqIO.parse(\"4-proteins.gb\",\"gb\"))\n",
    "\n",
    "# For every sequence in the file\n",
    "for s in seq:\n",
    "    print()\n",
    "    \n",
    "    # Protein/sequence name\n",
    "    name = s.name\n",
    "    print(name)\n",
    "    \n",
    "    # Finds all of the locations of the residues\n",
    "    for f in s.features:\n",
    "        # Check if the feature in genbank file is 'mutation'\n",
    "        if f.type == \"mutation\":\n",
    "            # Get the location that is not similar to consesnus when looking for a similarity between it and the aligned original sequence\n",
    "            print(f.location)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Different from consensus sequence: \n",
      "\n",
      "mRNA_1\n",
      "Position calculation starts from 0. In file it starts from 1.\n",
      "(1, 10)\n",
      "(12, 20)\n",
      "(25, 25)\n",
      "(36, 37)\n",
      "(57, 57)\n",
      "(60, 60)\n",
      "(78, 79)\n",
      "(102, 102)\n",
      "(109, 109)\n",
      "(111, 123)\n",
      "(128, 129)\n",
      "(132, 133)\n",
      "(135, 136)\n",
      "(138, 138)\n",
      "(140, 140)\n",
      "(143, 145)\n",
      "(153, 154)\n",
      "(156, 156)\n",
      "(158, 158)\n",
      "(160, 160)\n",
      "(162, 162)\n",
      "(164, 166)\n",
      "(169, 169)\n",
      "(171, 171)\n",
      "(175, 175)\n",
      "(178, 180)\n",
      "(182, 184)\n",
      "(188, 188)\n",
      "(190, 190)\n",
      "(199, 199)\n",
      "(203, 209)\n",
      "(217, 217)\n",
      "(225, 227)\n",
      "(229, 234)\n",
      "(236, 239)\n",
      "(241, 250)\n",
      "(254, 257)\n",
      "(259, 259)\n",
      "(271, 271)\n",
      "(281, 282)\n",
      "(291, 291)\n",
      "(294, 294)\n",
      "(296, 299)\n",
      "(304, 317)\n",
      "(320, 325)\n",
      "(366, 366)\n",
      "(429, 429)\n",
      "(438, 438)\n",
      "(447, 447)\n",
      "(450, 450)\n",
      "(455, 455)\n",
      "(470, 470)\n",
      "(476, 476)\n",
      "(485, 485)\n",
      "(488, 488)\n",
      "(494, 494)\n",
      "(501, 501)\n",
      "(512, 513)\n",
      "(524, 524)\n",
      "(531, 531)\n",
      "(535, 535)\n",
      "(539, 540)\n",
      "(546, 546)\n",
      "(548, 548)\n",
      "(555, 556)\n",
      "(559, 560)\n",
      "(562, 569)\n",
      "(571, 579)\n",
      "(581, 591)\n",
      "(593, 606)\n",
      "(608, 609)\n",
      "(611, 615)\n",
      "\n",
      "mRNA_2\n",
      "N/A\n",
      "\n",
      "mRNA_3\n",
      "Position calculation starts from 0. In file it starts from 1.\n",
      "(2, 2)\n",
      "(7, 7)\n",
      "(24, 25)\n",
      "(29, 30)\n",
      "(35, 35)\n",
      "(50, 50)\n",
      "(54, 54)\n",
      "(58, 58)\n",
      "(65, 65)\n",
      "(68, 68)\n",
      "(72, 72)\n",
      "(75, 75)\n",
      "(82, 88)\n",
      "(94, 94)\n",
      "(101, 101)\n",
      "(104, 104)\n",
      "(107, 109)\n",
      "(111, 113)\n",
      "(115, 116)\n",
      "(126, 126)\n",
      "(128, 128)\n",
      "(132, 133)\n",
      "(138, 138)\n",
      "(141, 143)\n",
      "(146, 147)\n",
      "(150, 151)\n",
      "(153, 153)\n",
      "(158, 159)\n",
      "(162, 162)\n",
      "(167, 168)\n",
      "(172, 172)\n",
      "(174, 177)\n",
      "(180, 181)\n",
      "(183, 186)\n",
      "(188, 188)\n",
      "(190, 190)\n",
      "(192, 192)\n",
      "(194, 194)\n",
      "(198, 199)\n",
      "(201, 201)\n",
      "(203, 203)\n",
      "(210, 212)\n",
      "(218, 218)\n",
      "(225, 225)\n",
      "(230, 230)\n",
      "(232, 232)\n",
      "(234, 234)\n",
      "(236, 236)\n",
      "(238, 238)\n",
      "(240, 240)\n",
      "(242, 242)\n",
      "(249, 249)\n",
      "(251, 251)\n",
      "(256, 256)\n",
      "(258, 258)\n",
      "(266, 267)\n",
      "(269, 269)\n",
      "(271, 271)\n",
      "(274, 279)\n",
      "(289, 289)\n",
      "(291, 291)\n",
      "(293, 293)\n",
      "(297, 297)\n",
      "(299, 300)\n",
      "(305, 307)\n",
      "(312, 313)\n",
      "(317, 318)\n",
      "(320, 320)\n",
      "(323, 323)\n",
      "(329, 329)\n",
      "(334, 334)\n",
      "(338, 339)\n",
      "(345, 345)\n",
      "(347, 347)\n",
      "\n",
      "mRNA_4\n",
      "Position calculation starts from 0. In file it starts from 1.\n",
      "(1, 31)\n",
      "(33, 40)\n",
      "(42, 55)\n",
      "(57, 57)\n",
      "(59, 62)\n",
      "(65, 65)\n",
      "(67, 67)\n",
      "(69, 69)\n",
      "(71, 75)\n",
      "(78, 78)\n",
      "(81, 81)\n",
      "(84, 84)\n",
      "(86, 88)\n",
      "(90, 91)\n",
      "(93, 108)\n",
      "(110, 146)\n",
      "(148, 153)\n",
      "(155, 167)\n",
      "(169, 172)\n",
      "(175, 178)\n",
      "(180, 188)\n",
      "(191, 201)\n",
      "(203, 205)\n",
      "(207, 208)\n",
      "(210, 211)\n",
      "(213, 214)\n",
      "(217, 227)\n",
      "(229, 236)\n",
      "(239, 240)\n",
      "(243, 280)\n",
      "(282, 282)\n",
      "(285, 285)\n",
      "(287, 289)\n",
      "(291, 292)\n",
      "(295, 296)\n",
      "(298, 299)\n",
      "(301, 303)\n",
      "(305, 308)\n",
      "(310, 313)\n",
      "(315, 317)\n",
      "(320, 320)\n",
      "(322, 322)\n",
      "(324, 326)\n",
      "(328, 329)\n",
      "(331, 331)\n",
      "(333, 335)\n",
      "(339, 339)\n",
      "(341, 342)\n",
      "(344, 345)\n",
      "(349, 350)\n",
      "(352, 352)\n",
      "(354, 358)\n",
      "(360, 361)\n",
      "(366, 367)\n",
      "(369, 375)\n",
      "(377, 379)\n",
      "(383, 391)\n",
      "(393, 395)\n",
      "(397, 398)\n",
      "(400, 401)\n",
      "(403, 410)\n",
      "(414, 415)\n",
      "(418, 423)\n",
      "(425, 426)\n",
      "(428, 433)\n",
      "(436, 436)\n",
      "(438, 438)\n",
      "(440, 440)\n",
      "(443, 443)\n",
      "(445, 451)\n",
      "(455, 458)\n",
      "(460, 460)\n",
      "(462, 465)\n",
      "(467, 476)\n",
      "(478, 481)\n",
      "(484, 487)\n",
      "(490, 495)\n",
      "(498, 500)\n",
      "(502, 510)\n",
      "(512, 515)\n",
      "(517, 522)\n",
      "(524, 528)\n",
      "(530, 531)\n",
      "(533, 536)\n",
      "(538, 540)\n",
      "(543, 546)\n",
      "(554, 555)\n",
      "(558, 558)\n",
      "(560, 573)\n",
      "(577, 581)\n",
      "(583, 600)\n",
      "(602, 623)\n",
      "(625, 627)\n",
      "(629, 630)\n",
      "(632, 642)\n",
      "(644, 650)\n",
      "(657, 657)\n",
      "(659, 661)\n",
      "(663, 691)\n",
      "(693, 707)\n",
      "(709, 745)\n",
      "(747, 755)\n",
      "(757, 758)\n",
      "(760, 767)\n",
      "(771, 772)\n",
      "(774, 785)\n",
      "(787, 795)\n",
      "(798, 802)\n",
      "(804, 805)\n",
      "(807, 807)\n",
      "(809, 810)\n",
      "(812, 813)\n",
      "(815, 827)\n",
      "(829, 842)\n",
      "(844, 860)\n",
      "(862, 865)\n",
      "(867, 878)\n",
      "(880, 881)\n",
      "(883, 884)\n",
      "(889, 894)\n",
      "(896, 897)\n",
      "(900, 901)\n",
      "(903, 912)\n",
      "(914, 936)\n",
      "(938, 964)\n",
      "(966, 994)\n",
      "\n",
      "mRNA_5\n",
      "Position calculation starts from 0. In file it starts from 1.\n",
      "(7, 8)\n",
      "(12, 18)\n",
      "(20, 22)\n",
      "(26, 29)\n",
      "(32, 32)\n",
      "(34, 34)\n",
      "(36, 36)\n",
      "(40, 40)\n",
      "(42, 42)\n",
      "(44, 44)\n",
      "(48, 48)\n",
      "(50, 51)\n",
      "(55, 56)\n",
      "(59, 59)\n",
      "(62, 62)\n",
      "(64, 65)\n",
      "(68, 69)\n",
      "(72, 72)\n",
      "(86, 89)\n",
      "(94, 94)\n",
      "(108, 108)\n",
      "(112, 112)\n",
      "(115, 115)\n",
      "(120, 120)\n",
      "(123, 124)\n",
      "(128, 128)\n",
      "(135, 136)\n",
      "(138, 138)\n",
      "(144, 144)\n",
      "(148, 148)\n",
      "(156, 157)\n",
      "(167, 168)\n",
      "(170, 170)\n",
      "(172, 172)\n",
      "(178, 178)\n",
      "(184, 185)\n",
      "(195, 195)\n",
      "(199, 199)\n",
      "(201, 202)\n",
      "(205, 205)\n",
      "(215, 215)\n",
      "(223, 223)\n",
      "(225, 225)\n",
      "(229, 229)\n",
      "(234, 234)\n",
      "(241, 241)\n",
      "(247, 248)\n",
      "(251, 251)\n",
      "(254, 256)\n",
      "(260, 260)\n",
      "(264, 265)\n",
      "(272, 274)\n",
      "(277, 277)\n",
      "(279, 279)\n",
      "(281, 282)\n",
      "(285, 285)\n",
      "(288, 290)\n",
      "(292, 292)\n",
      "(295, 295)\n",
      "(297, 297)\n",
      "(299, 300)\n",
      "(302, 303)\n",
      "(315, 315)\n",
      "(317, 317)\n",
      "(322, 322)\n",
      "(324, 324)\n",
      "(327, 327)\n",
      "(330, 331)\n",
      "(334, 335)\n",
      "(340, 340)\n",
      "(345, 346)\n",
      "(351, 352)\n",
      "(354, 358)\n",
      "(362, 363)\n",
      "(366, 367)\n",
      "(369, 369)\n",
      "(371, 371)\n",
      "(374, 377)\n",
      "(379, 380)\n",
      "(382, 382)\n",
      "(388, 389)\n",
      "(392, 392)\n",
      "(399, 399)\n",
      "\n",
      "mRNA_6\n",
      "Position calculation starts from 0. In file it starts from 1.\n",
      "(4, 4)\n",
      "(14, 14)\n",
      "(31, 35)\n",
      "(39, 39)\n",
      "(42, 43)\n",
      "(47, 47)\n",
      "(60, 60)\n",
      "(68, 71)\n",
      "(75, 75)\n",
      "(84, 84)\n",
      "(87, 87)\n",
      "(91, 91)\n",
      "(94, 94)\n",
      "(99, 100)\n",
      "(104, 104)\n",
      "(107, 108)\n",
      "(112, 112)\n",
      "(115, 115)\n",
      "(118, 118)\n",
      "(120, 120)\n",
      "(123, 123)\n",
      "(125, 125)\n",
      "(130, 130)\n",
      "(133, 135)\n",
      "(137, 137)\n",
      "(141, 142)\n",
      "(147, 147)\n",
      "(150, 152)\n",
      "(167, 167)\n",
      "(171, 171)\n",
      "(174, 176)\n",
      "(181, 181)\n",
      "(197, 197)\n",
      "(199, 200)\n",
      "(202, 205)\n",
      "(207, 209)\n",
      "(213, 214)\n",
      "(217, 217)\n",
      "(230, 230)\n",
      "(234, 235)\n",
      "(237, 237)\n",
      "(239, 239)\n",
      "(241, 241)\n",
      "(245, 247)\n",
      "(250, 251)\n",
      "(253, 256)\n",
      "(258, 258)\n",
      "(262, 262)\n",
      "(265, 265)\n",
      "(272, 272)\n",
      "(274, 276)\n",
      "(287, 292)\n",
      "(295, 295)\n",
      "(298, 298)\n",
      "(300, 300)\n",
      "(307, 307)\n",
      "(311, 312)\n",
      "(316, 317)\n",
      "(327, 328)\n",
      "(331, 331)\n",
      "(336, 337)\n",
      "(342, 342)\n",
      "(353, 353)\n",
      "(355, 356)\n",
      "(363, 363)\n",
      "(365, 366)\n",
      "(381, 381)\n",
      "(386, 386)\n",
      "(388, 390)\n",
      "(392, 393)\n",
      "(396, 396)\n",
      "(403, 408)\n",
      "(415, 417)\n",
      "(419, 419)\n",
      "(422, 422)\n",
      "(428, 428)\n",
      "(431, 431)\n",
      "(444, 445)\n",
      "(447, 447)\n",
      "(450, 454)\n",
      "(456, 457)\n",
      "(461, 461)\n",
      "(463, 464)\n",
      "(468, 470)\n",
      "(472, 474)\n",
      "(476, 478)\n",
      "(483, 486)\n",
      "(490, 493)\n",
      "(495, 495)\n",
      "(498, 498)\n",
      "(502, 503)\n",
      "(505, 505)\n",
      "(507, 507)\n",
      "(509, 509)\n",
      "(511, 511)\n",
      "(514, 514)\n",
      "(517, 518)\n",
      "(523, 523)\n",
      "(525, 526)\n",
      "(534, 535)\n",
      "(540, 542)\n",
      "(550, 551)\n",
      "(557, 557)\n",
      "(565, 566)\n",
      "(570, 570)\n",
      "(572, 572)\n",
      "(574, 574)\n",
      "(579, 579)\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2. COMPARED WITH THE ORIGINAL UNALIGNED SEQUENCE AND SAVE RESIDUES IN FILE\n",
    "\n",
    "\n",
    "from Bio import AlignIO\n",
    "from Bio import SeqIO\n",
    "from Bio.Align import AlignInfo # To generate a trivial 'dumb' consensus\n",
    "from Bio.Align.AlignInfo import SummaryInfo\n",
    "## load AlignInfo modules and SummaryInfo submodules\n",
    "import Bio.SeqFeature as BSF\n",
    "\n",
    "try:\n",
    "    from itertools import izip as zip\n",
    "except ImportError: # will be 3.x series\n",
    "    pass\n",
    "\n",
    "# Putting all of the genbank file into the list\n",
    "proteins3 = list(SeqIO.parse(\"3-proteins.gb\",\"gb\"))\n",
    "\n",
    "print(\"Different from consensus sequence: \\n\")\n",
    "\n",
    "for protein in proteins3:\n",
    "    \n",
    "    # Name of the protein\n",
    "    name = protein.id\n",
    "    print(name)\n",
    "    if protein.id == \"mRNA_2\":\n",
    "        print(\"N/A\")\n",
    "    else:\n",
    "        print(\"Position calculation starts from 0. In file it starts from 1.\")\n",
    "    \n",
    "    # Protein sequence\n",
    "    seq = protein.seq\n",
    "    #print(seq)\n",
    "    \n",
    "    # Alignment files being read base on protein name\n",
    "    align = AlignIO.read(f'3-alignment-{name}.fasta', \"fasta\")\n",
    "    #print(align)\n",
    "\n",
    "    # Summary of the alignment\n",
    "    summary = SummaryInfo(align)\n",
    "\n",
    "    # Create dumb consensus sequence\n",
    "    dumb_cons = summary.dumb_consensus(threshold=0.6)\n",
    "\n",
    "    index = 0\n",
    "    \n",
    "    # Comparison of dumb consensus sequence with out target protein\n",
    "    # Results are our positions that do not match\n",
    "    for n1, n2 in zip(seq,dumb_cons):\n",
    "        \n",
    "        if n1 != n2 and n1 not in \"-X\" and n2 not in \"-X\":\n",
    "                \n",
    "            # Creation of GO feature for our protein sequences'\n",
    "            # Len vieno protein o ne visu proteins yra tas 1..616 salia BLAST\n",
    "            feature = BSF.SeqFeature(BSF.FeatureLocation(index,index + 1), type='mutation')\n",
    "\n",
    "            # We add a qualifier 'db_xref' to the GO IDs (Terms) and add it to BLAST feature\n",
    "            feature.qualifiers['observed'] = n1\n",
    "\n",
    "            # In the same GO feature we add location/qualifier that is 'evidence'\n",
    "            feature.qualifiers['consensus'] = n2\n",
    "\n",
    "            # Then for every protein (not the whole stack of proteins but just every individual protein/mRNA)in the genbank file \n",
    "            # we add those features that have a name GO on the left collum\n",
    "            # and we add all of the feature items, I mean all of those ids with the qualifier\n",
    "            protein.features.append(feature)\n",
    "     \n",
    "        if n1 != \"X\":\n",
    "            index += 1\n",
    "     \n",
    "    # The [] snipit is from https://stackoverflow.com/questions/8545492/find-the-position-of-difference-between-two-strings \n",
    "    # by ovgolovin\n",
    "    # modified to ignore ambiguous character X\n",
    "    results = [i for i, (ch1, ch2) in enumerate(zip(dumb_cons, seq)) if ch1 != ch2 and ch1 != \"X\" and ch2 not in \"X\"]\n",
    "    #print(name, results)\n",
    "\n",
    "    # The start of our gourped positions\n",
    "    start = None\n",
    "    \n",
    "    # Grouped position; Our positions that are not similar to the consensus\n",
    "    # Basically [(1,10),(12,20),...] and so on\n",
    "    grouped = []\n",
    "    \n",
    "    # For every index of our positions\n",
    "    for i, pos in enumerate(results):\n",
    "        # If we havent started than\n",
    "        if start is None:\n",
    "            # Make our start the postion\n",
    "            start = pos\n",
    "            \n",
    "            # Check so it would start at then of our postion list,\n",
    "            # if not it would an error\n",
    "            if len(results) == i + 1:\n",
    "                grouped.append((start, start))\n",
    "                break\n",
    "\n",
    "        # Check if next element exists\n",
    "        if len(results) == i + 1:\n",
    "            grouped.append((start, pos))\n",
    "            break\n",
    "        \n",
    "        # If the position group has to same numbers like for example (7,7)\n",
    "        # It means that its that Nth position, or in the examples case, the position is just 7\n",
    "        next_pos = results[i + 1]\n",
    "        if pos + 1 != next_pos:\n",
    "            grouped.append((start, pos))\n",
    "            start = None\n",
    "    \n",
    "    for position in grouped:\n",
    "        print(position)\n",
    "    print()\n",
    "                \n",
    "# Write to genbank file\n",
    "SeqIO.write(proteins3, \"4-proteins2.gb\",\"gb\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I did both ways, by comparing the original sequence which was aligned (OA) with the consensus (gapped) and also the original unaligned (OU) sequence with the consensus (dumb).<p>\n",
    "I have noticed that the OU sequences compared with consensus has a lot of different residues. Most of the residues do not match the consensus. However, while comparing the OA with the consensus it had less difference bewteen them. I think that it is because, for example we are comparing our raw sequnce with the ungapped consensus and it will deffinetaly have a lot of differences exluding the ambiguous characters X. While using OA and gaped consensus it has less differences because they are a bit more well aligned together, because of the gaps in both of the consensus and OA sequences. I think that the gapped consensus keeps the aligments (all the gaps and all)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
